// twilio/lib/rest/taskrouter/v1/workspace/task/reservation.js
'use strict';

/* jshint ignore:start */
/**
 * This code was generated by
 * \ / _    _  _|   _  _
 *  | (_)\/(_)(_|\/| |(/_  v1.0.0
 *       /       /
 */
/* jshint ignore:end */

var Q = require('q');  /* jshint ignore:line */
var _ = require('lodash');  /* jshint ignore:line */
var Page = require('../../../../../base/Page');  /* jshint ignore:line */
var deserialize = require(
    '../../../../../base/deserialize');  /* jshint ignore:line */
var serialize = require(
    '../../../../../base/serialize');  /* jshint ignore:line */
var values = require('../../../../../base/values');  /* jshint ignore:line */

var ReservationList;
var ReservationPage;
var ReservationInstance;
var ReservationContext;

/* jshint ignore:start */
/**
 * @description Initialize the ReservationList
 *
 * @param {Twilio.Taskrouter.V1} version - Version of the resource
 * @param {string} workspaceSid -
 *          The ID of the Workspace that this task is contained within.
 * @param {string} taskSid - The ID of the reserved Task
 */
/* jshint ignore:end */
ReservationList = function ReservationList(version, workspaceSid, taskSid) {
  /* jshint ignore:start */
  /**
   * @param {string} sid - sid of instance
   *
   * @returns {Twilio.Taskrouter.V1.WorkspaceContext.TaskContext.ReservationContext}
   */
  /* jshint ignore:end */
  function ReservationListInstance(sid) {
    return ReservationListInstance.get(sid);
  }

  ReservationListInstance._version = version;
  // Path Solution
  ReservationListInstance._solution = {workspaceSid: workspaceSid, taskSid: taskSid};
  ReservationListInstance._uri = _.template(
    '/Workspaces/<%= workspaceSid %>/Tasks/<%= taskSid %>/Reservations' // jshint ignore:line
  )(ReservationListInstance._solution);
  /* jshint ignore:start */
  /**
   * Streams ReservationInstance records from the API.
   *
   * This operation lazily loads records as efficiently as possible until the limit
   * is reached.
   *
   * The results are passed into the callback function, so this operation is memory efficient.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {reservation.status} [opts.reservationStatus] -
   *          Returns the list of reservations for a task with a specified ReservationStatus
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         each() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no pageSize is defined but a limit is defined,
   *         each() will attempt to read the limit with the most efficient
   *         page size, i.e. min(limit, 1000)
   * @param {Function} [opts.callback] -
   *         Function to process each record. If this and a positional
   *         callback are passed, this one will be used
   * @param {Function} [opts.done] -
   *          Function to be called upon completion of streaming
   * @param {Function} [callback] - Function to process each record
   */
  /* jshint ignore:end */
  ReservationListInstance.each = function each(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    if (opts.callback) {
      callback = opts.callback;
    }
    if (_.isUndefined(callback)) {
      throw new Error('Callback function must be provided');
    }

    var done = false;
    var currentPage = 1;
    var currentResource = 0;
    var limits = this._version.readLimits({
      limit: opts.limit,
      pageSize: opts.pageSize
    });

    function onComplete(error) {
      done = true;
      if (_.isFunction(opts.done)) {
        opts.done(error);
      }
    }

    function fetchNextPage(fn) {
      var promise = fn();
      if (_.isUndefined(promise)) {
        onComplete();
        return;
      }

      promise.then(function(page) {
        _.each(page.instances, function(instance) {
          if (done || (!_.isUndefined(opts.limit) && currentResource >= opts.limit)) {
            done = true;
            return false;
          }

          currentResource++;
          callback(instance, onComplete);
        });

        if ((limits.pageLimit && limits.pageLimit <= currentPage)) {
          onComplete();
        } else if (!done) {
          currentPage++;
          fetchNextPage(_.bind(page.nextPage, page));
        }
      });

      promise.catch(onComplete);
    }

    fetchNextPage(_.bind(this.page, this, _.merge(opts, limits)));
  };

  /* jshint ignore:start */
  /**
   * Lists ReservationInstance records from the API as a list.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {reservation.status} [opts.reservationStatus] -
   *          Returns the list of reservations for a task with a specified ReservationStatus
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         list() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no page_size is defined but a limit is defined,
   *         list() will attempt to read the limit with the most
   *         efficient page size, i.e. min(limit, 1000)
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  ReservationListInstance.list = function list(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    var deferred = Q.defer();
    var allResources = [];
    opts.callback = function(resource, done) {
      allResources.push(resource);

      if (!_.isUndefined(opts.limit) && allResources.length === opts.limit) {
        done();
      }
    };

    opts.done = function(error) {
      if (_.isUndefined(error)) {
        deferred.resolve(allResources);
      } else {
        deferred.reject(error);
      }
    };

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    this.each(opts);
    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single page of ReservationInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {reservation.status} [opts.reservationStatus] -
   *          Returns the list of reservations for a task with a specified ReservationStatus
   * @param {string} [opts.pageToken] - PageToken provided by the API
   * @param {number} [opts.pageNumber] -
   *          Page Number, this value is simply for client state
   * @param {number} [opts.pageSize] - Number of records to return, defaults to 50
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  ReservationListInstance.page = function page(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};

    var deferred = Q.defer();
    var data = values.of({
      'ReservationStatus': _.get(opts, 'reservationStatus'),
      'PageToken': opts.pageToken,
      'Page': opts.pageNumber,
      'PageSize': opts.pageSize
    });

    var promise = this._version.page({uri: this._uri, method: 'GET', params: data});

    promise = promise.then(function(payload) {
      deferred.resolve(new ReservationPage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single target page of ReservationInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {string} [targetUrl] - API-generated URL for the requested results page
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  ReservationListInstance.getPage = function getPage(targetUrl, callback) {
    var deferred = Q.defer();

    var promise = this._version._domain.twilio.request({method: 'GET', uri: targetUrl});

    promise = promise.then(function(payload) {
      deferred.resolve(new ReservationPage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Constructs a reservation
   *
   * @param {string} sid - The sid
   *
   * @returns {Twilio.Taskrouter.V1.WorkspaceContext.TaskContext.ReservationContext}
   */
  /* jshint ignore:end */
  ReservationListInstance.get = function get(sid) {
    return new ReservationContext(
      this._version,
      this._solution.workspaceSid,
      this._solution.taskSid,
      sid
    );
  };

  return ReservationListInstance;
};


/* jshint ignore:start */
/**
 * Initialize the ReservationPage
 *
 * @param {V1} version - Version of the resource
 * @param {Response<string>} response - Response from the API
 * @param {ReservationSolution} solution - Path solution
 *
 * @returns ReservationPage
 */
/* jshint ignore:end */
ReservationPage = function ReservationPage(version, response, solution) {
  // Path Solution
  this._solution = solution;

  Page.prototype.constructor.call(this, version, response, this._solution);
};

_.extend(ReservationPage.prototype, Page.prototype);
ReservationPage.prototype.constructor = ReservationPage;

/* jshint ignore:start */
/**
 * Build an instance of ReservationInstance
 *
 * @param {ReservationPayload} payload - Payload response from the API
 *
 * @returns ReservationInstance
 */
/* jshint ignore:end */
ReservationPage.prototype.getInstance = function getInstance(payload) {
  return new ReservationInstance(
    this._version,
    payload,
    this._solution.workspaceSid,
    this._solution.taskSid
  );
};


/* jshint ignore:start */
/**
 * Initialize the ReservationContext
 *
 * @property {string} accountSid - The ID of the Account that owns this Task
 * @property {Date} dateCreated - The date_created
 * @property {Date} dateUpdated - The date_updated
 * @property {reservation.status} reservationStatus -
 *          The current status of the reservation.
 * @property {string} sid - The unique ID of this Reservation.
 * @property {string} taskSid - The ID of the reserved Task
 * @property {string} workerName -
 *          Human readable description of the Worker that is reserved
 * @property {string} workerSid - The ID of the reserved Worker
 * @property {string} workspaceSid -
 *          The ID of the Workspace that this task is contained within.
 * @property {string} url - The url
 * @property {string} links - The links
 *
 * @param {V1} version - Version of the resource
 * @param {ReservationPayload} payload - The instance payload
 * @param {sid} workspaceSid -
 *          The ID of the Workspace that this task is contained within.
 * @param {sid} taskSid - The ID of the reserved Task
 * @param {sid} sid - The sid
 */
/* jshint ignore:end */
ReservationInstance = function ReservationInstance(version, payload,
                                                    workspaceSid, taskSid, sid)
                                                    {
  this._version = version;

  // Marshaled Properties
  this.accountSid = payload.account_sid; // jshint ignore:line
  this.dateCreated = deserialize.iso8601DateTime(payload.date_created); // jshint ignore:line
  this.dateUpdated = deserialize.iso8601DateTime(payload.date_updated); // jshint ignore:line
  this.reservationStatus = payload.reservation_status; // jshint ignore:line
  this.sid = payload.sid; // jshint ignore:line
  this.taskSid = payload.task_sid; // jshint ignore:line
  this.workerName = payload.worker_name; // jshint ignore:line
  this.workerSid = payload.worker_sid; // jshint ignore:line
  this.workspaceSid = payload.workspace_sid; // jshint ignore:line
  this.url = payload.url; // jshint ignore:line
  this.links = payload.links; // jshint ignore:line

  // Context
  this._context = undefined;
  this._solution = {workspaceSid: workspaceSid, taskSid: taskSid, sid: sid || this.sid, };
};

Object.defineProperty(ReservationInstance.prototype,
  '_proxy', {
  get: function() {
    if (!this._context) {
      this._context = new ReservationContext(
        this._version,
        this._solution.workspaceSid,
        this._solution.taskSid,
        this._solution.sid
      );
    }

    return this._context;
  }
});

/* jshint ignore:start */
/**
 * fetch a ReservationInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ReservationInstance
 */
/* jshint ignore:end */
ReservationInstance.prototype.fetch = function fetch(callback) {
  return this._proxy.fetch(callback);
};

/* jshint ignore:start */
/**
 * update a ReservationInstance
 *
 * @param {object} [opts] - Options for request
 * @param {reservation.status} [opts.reservationStatus] - New reservation status
 * @param {string} [opts.workerActivitySid] -
 *          New worker activity sid if rejecting a reservation
 * @param {string} [opts.instruction] - Assignment instruction for reservation
 * @param {string} [opts.dequeuePostWorkActivitySid] -
 *          New worker activity sid after executing a Dequeue instruction
 * @param {string} [opts.dequeueFrom] -
 *          Caller ID for the call to the worker when executing a Dequeue instruction
 * @param {string} [opts.dequeueRecord] -
 *          Attribute to record both legs of a call when executing a Dequeue instruction
 * @param {number} [opts.dequeueTimeout] -
 *          Timeout for call when executing a Dequeue instruction
 * @param {string} [opts.dequeueTo] -
 *          Contact URI of the worker when executing a Dequeue instruction
 * @param {string} [opts.dequeueStatusCallbackUrl] -
 *          Callback URL for completed call event when executing a Dequeue instruction
 * @param {string} [opts.callFrom] -
 *          Caller ID for the outbound call when executing a Call instruction
 * @param {string} [opts.callRecord] -
 *          Attribute to record both legs of a call when executing a Call instruction
 * @param {number} [opts.callTimeout] -
 *          Timeout for call when executing a Call instruction
 * @param {string} [opts.callTo] -
 *          Contact URI of the worker when executing a Call instruction
 * @param {string} [opts.callUrl] -
 *          TwiML URI executed on answering the worker's leg as a result of the Call instruction
 * @param {string} [opts.callStatusCallbackUrl] -
 *          Callback URL for completed call event when executing a Call instruction
 * @param {boolean} [opts.callAccept] -
 *          Flag to determine if reservation should be accepted when executing a Call instruction
 * @param {string} [opts.redirectCallSid] -
 *          Call sid of the call parked in the queue when executing a Redirect instruction
 * @param {boolean} [opts.redirectAccept] -
 *          Flag to determine if reservation should be accepted when executing a Redirect instruction
 * @param {string} [opts.redirectUrl] -
 *          TwiML URI to redirect the call to when executing the Redirect instruction
 * @param {string} [opts.to] -
 *          Contact URI of the worker when executing a Conference instruction
 * @param {string} [opts.from] -
 *          Caller ID for the call to the worker when executing a Conference instruction
 * @param {string} [opts.statusCallback] - The status_callback
 * @param {string} [opts.statusCallbackMethod] - The status_callback_method
 * @param {reservation.call_status|list} [opts.statusCallbackEvent] -
 *          The status_callback_event
 * @param {number} [opts.timeout] -
 *          Timeout for call when executing a Conference instruction
 * @param {boolean} [opts.record] - The record
 * @param {boolean} [opts.muted] - The muted
 * @param {string} [opts.beep] - The beep
 * @param {boolean} [opts.startConferenceOnEnter] - The start_conference_on_enter
 * @param {boolean} [opts.endConferenceOnExit] - The end_conference_on_exit
 * @param {string} [opts.waitUrl] - The wait_url
 * @param {string} [opts.waitMethod] - The wait_method
 * @param {boolean} [opts.earlyMedia] - The early_media
 * @param {number} [opts.maxParticipants] - The max_participants
 * @param {string} [opts.conferenceStatusCallback] - The conference_status_callback
 * @param {string} [opts.conferenceStatusCallbackMethod] -
 *          The conference_status_callback_method
 * @param {reservation.conference_event|list} [opts.conferenceStatusCallbackEvent] -
 *          The conference_status_callback_event
 * @param {string} [opts.conferenceRecord] - The conference_record
 * @param {string} [opts.conferenceTrim] - The conference_trim
 * @param {string} [opts.recordingChannels] - The recording_channels
 * @param {string} [opts.recordingStatusCallback] - The recording_status_callback
 * @param {string} [opts.recordingStatusCallbackMethod] -
 *          The recording_status_callback_method
 * @param {string} [opts.conferenceRecordingStatusCallback] -
 *          The conference_recording_status_callback
 * @param {string} [opts.conferenceRecordingStatusCallbackMethod] -
 *          The conference_recording_status_callback_method
 * @param {string} [opts.region] - The region
 * @param {string} [opts.sipAuthUsername] - The sip_auth_username
 * @param {string} [opts.sipAuthPassword] - The sip_auth_password
 * @param {string|list} [opts.dequeueStatusCallbackEvent] -
 *          Call progress events sent via webhooks as a result of a Dequeue instruction
 * @param {string} [opts.postWorkActivitySid] -
 *          New worker activity sid after executing a Conference instruction
 * @param {reservation.supervisor_mode} [opts.supervisorMode] -
 *          Supervisor mode when executing the Supervise instruction
 * @param {string} [opts.supervisor] -
 *          Supervisor sid/uri when executing the Supervise instruction
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ReservationInstance
 */
/* jshint ignore:end */
ReservationInstance.prototype.update = function update(opts, callback) {
  return this._proxy.update(opts, callback);
};

/* jshint ignore:start */
/**
 * Produce a plain JSON object version of the ReservationInstance for serialization.
 * Removes any circular references in the object.
 *
 * @returns Object
 */
/* jshint ignore:end */
ReservationInstance.prototype.toJSON = function toJSON() {
  let clone = {};
  _.forOwn(this, function(value, key) {
    if (!_.startsWith(key, '_') && ! _.isFunction(value)) {
      clone[key] = value;
    }
  });
  return clone;
};


/* jshint ignore:start */
/**
 * Initialize the ReservationContext
 *
 * @param {V1} version - Version of the resource
 * @param {sid} workspaceSid - The workspace_sid
 * @param {sid} taskSid - The task_sid
 * @param {sid} sid - The sid
 */
/* jshint ignore:end */
ReservationContext = function ReservationContext(version, workspaceSid, taskSid,
                                                  sid) {
  this._version = version;

  // Path Solution
  this._solution = {workspaceSid: workspaceSid, taskSid: taskSid, sid: sid, };
  this._uri = _.template(
    '/Workspaces/<%= workspaceSid %>/Tasks/<%= taskSid %>/Reservations/<%= sid %>' // jshint ignore:line
  )(this._solution);
};

/* jshint ignore:start */
/**
 * fetch a ReservationInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ReservationInstance
 */
/* jshint ignore:end */
ReservationContext.prototype.fetch = function fetch(callback) {
  var deferred = Q.defer();
  var promise = this._version.fetch({uri: this._uri, method: 'GET'});

  promise = promise.then(function(payload) {
    deferred.resolve(new ReservationInstance(
      this._version,
      payload,
      this._solution.workspaceSid,
      this._solution.taskSid,
      this._solution.sid
    ));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

/* jshint ignore:start */
/**
 * update a ReservationInstance
 *
 * @param {object} [opts] - Options for request
 * @param {reservation.status} [opts.reservationStatus] - New reservation status
 * @param {string} [opts.workerActivitySid] -
 *          New worker activity sid if rejecting a reservation
 * @param {string} [opts.instruction] - Assignment instruction for reservation
 * @param {string} [opts.dequeuePostWorkActivitySid] -
 *          New worker activity sid after executing a Dequeue instruction
 * @param {string} [opts.dequeueFrom] -
 *          Caller ID for the call to the worker when executing a Dequeue instruction
 * @param {string} [opts.dequeueRecord] -
 *          Attribute to record both legs of a call when executing a Dequeue instruction
 * @param {number} [opts.dequeueTimeout] -
 *          Timeout for call when executing a Dequeue instruction
 * @param {string} [opts.dequeueTo] -
 *          Contact URI of the worker when executing a Dequeue instruction
 * @param {string} [opts.dequeueStatusCallbackUrl] -
 *          Callback URL for completed call event when executing a Dequeue instruction
 * @param {string} [opts.callFrom] -
 *          Caller ID for the outbound call when executing a Call instruction
 * @param {string} [opts.callRecord] -
 *          Attribute to record both legs of a call when executing a Call instruction
 * @param {number} [opts.callTimeout] -
 *          Timeout for call when executing a Call instruction
 * @param {string} [opts.callTo] -
 *          Contact URI of the worker when executing a Call instruction
 * @param {string} [opts.callUrl] -
 *          TwiML URI executed on answering the worker's leg as a result of the Call instruction
 * @param {string} [opts.callStatusCallbackUrl] -
 *          Callback URL for completed call event when executing a Call instruction
 * @param {boolean} [opts.callAccept] -
 *          Flag to determine if reservation should be accepted when executing a Call instruction
 * @param {string} [opts.redirectCallSid] -
 *          Call sid of the call parked in the queue when executing a Redirect instruction
 * @param {boolean} [opts.redirectAccept] -
 *          Flag to determine if reservation should be accepted when executing a Redirect instruction
 * @param {string} [opts.redirectUrl] -
 *          TwiML URI to redirect the call to when executing the Redirect instruction
 * @param {string} [opts.to] -
 *          Contact URI of the worker when executing a Conference instruction
 * @param {string} [opts.from] -
 *          Caller ID for the call to the worker when executing a Conference instruction
 * @param {string} [opts.statusCallback] - The status_callback
 * @param {string} [opts.statusCallbackMethod] - The status_callback_method
 * @param {reservation.call_status|list} [opts.statusCallbackEvent] -
 *          The status_callback_event
 * @param {number} [opts.timeout] -
 *          Timeout for call when executing a Conference instruction
 * @param {boolean} [opts.record] - The record
 * @param {boolean} [opts.muted] - The muted
 * @param {string} [opts.beep] - The beep
 * @param {boolean} [opts.startConferenceOnEnter] - The start_conference_on_enter
 * @param {boolean} [opts.endConferenceOnExit] - The end_conference_on_exit
 * @param {string} [opts.waitUrl] - The wait_url
 * @param {string} [opts.waitMethod] - The wait_method
 * @param {boolean} [opts.earlyMedia] - The early_media
 * @param {number} [opts.maxParticipants] - The max_participants
 * @param {string} [opts.conferenceStatusCallback] - The conference_status_callback
 * @param {string} [opts.conferenceStatusCallbackMethod] -
 *          The conference_status_callback_method
 * @param {reservation.conference_event|list} [opts.conferenceStatusCallbackEvent] -
 *          The conference_status_callback_event
 * @param {string} [opts.conferenceRecord] - The conference_record
 * @param {string} [opts.conferenceTrim] - The conference_trim
 * @param {string} [opts.recordingChannels] - The recording_channels
 * @param {string} [opts.recordingStatusCallback] - The recording_status_callback
 * @param {string} [opts.recordingStatusCallbackMethod] -
 *          The recording_status_callback_method
 * @param {string} [opts.conferenceRecordingStatusCallback] -
 *          The conference_recording_status_callback
 * @param {string} [opts.conferenceRecordingStatusCallbackMethod] -
 *          The conference_recording_status_callback_method
 * @param {string} [opts.region] - The region
 * @param {string} [opts.sipAuthUsername] - The sip_auth_username
 * @param {string} [opts.sipAuthPassword] - The sip_auth_password
 * @param {string|list} [opts.dequeueStatusCallbackEvent] -
 *          Call progress events sent via webhooks as a result of a Dequeue instruction
 * @param {string} [opts.postWorkActivitySid] -
 *          New worker activity sid after executing a Conference instruction
 * @param {reservation.supervisor_mode} [opts.supervisorMode] -
 *          Supervisor mode when executing the Supervise instruction
 * @param {string} [opts.supervisor] -
 *          Supervisor sid/uri when executing the Supervise instruction
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ReservationInstance
 */
/* jshint ignore:end */
ReservationContext.prototype.update = function update(opts, callback) {
  if (_.isFunction(opts)) {
    callback = opts;
    opts = {};
  }
  opts = opts || {};

  var deferred = Q.defer();
  var data = values.of({
    'ReservationStatus': _.get(opts, 'reservationStatus'),
    'WorkerActivitySid': _.get(opts, 'workerActivitySid'),
    'Instruction': _.get(opts, 'instruction'),
    'DequeuePostWorkActivitySid': _.get(opts, 'dequeuePostWorkActivitySid'),
    'DequeueFrom': _.get(opts, 'dequeueFrom'),
    'DequeueRecord': _.get(opts, 'dequeueRecord'),
    'DequeueTimeout': _.get(opts, 'dequeueTimeout'),
    'DequeueTo': _.get(opts, 'dequeueTo'),
    'DequeueStatusCallbackUrl': _.get(opts, 'dequeueStatusCallbackUrl'),
    'CallFrom': _.get(opts, 'callFrom'),
    'CallRecord': _.get(opts, 'callRecord'),
    'CallTimeout': _.get(opts, 'callTimeout'),
    'CallTo': _.get(opts, 'callTo'),
    'CallUrl': _.get(opts, 'callUrl'),
    'CallStatusCallbackUrl': _.get(opts, 'callStatusCallbackUrl'),
    'CallAccept': serialize.bool(_.get(opts, 'callAccept')),
    'RedirectCallSid': _.get(opts, 'redirectCallSid'),
    'RedirectAccept': serialize.bool(_.get(opts, 'redirectAccept')),
    'RedirectUrl': _.get(opts, 'redirectUrl'),
    'To': _.get(opts, 'to'),
    'From': _.get(opts, 'from'),
    'StatusCallback': _.get(opts, 'statusCallback'),
    'StatusCallbackMethod': _.get(opts, 'statusCallbackMethod'),
    'StatusCallbackEvent': serialize.map(_.get(opts, 'statusCallbackEvent'), function(e) { return e; }),
    'Timeout': _.get(opts, 'timeout'),
    'Record': serialize.bool(_.get(opts, 'record')),
    'Muted': serialize.bool(_.get(opts, 'muted')),
    'Beep': _.get(opts, 'beep'),
    'StartConferenceOnEnter': serialize.bool(_.get(opts, 'startConferenceOnEnter')),
    'EndConferenceOnExit': serialize.bool(_.get(opts, 'endConferenceOnExit')),
    'WaitUrl': _.get(opts, 'waitUrl'),
    'WaitMethod': _.get(opts, 'waitMethod'),
    'EarlyMedia': serialize.bool(_.get(opts, 'earlyMedia')),
    'MaxParticipants': _.get(opts, 'maxParticipants'),
    'ConferenceStatusCallback': _.get(opts, 'conferenceStatusCallback'),
    'ConferenceStatusCallbackMethod': _.get(opts, 'conferenceStatusCallbackMethod'),
    'ConferenceStatusCallbackEvent': serialize.map(_.get(opts, 'conferenceStatusCallbackEvent'), function(e) { return e; }),
    'ConferenceRecord': _.get(opts, 'conferenceRecord'),
    'ConferenceTrim': _.get(opts, 'conferenceTrim'),
    'RecordingChannels': _.get(opts, 'recordingChannels'),
    'RecordingStatusCallback': _.get(opts, 'recordingStatusCallback'),
    'RecordingStatusCallbackMethod': _.get(opts, 'recordingStatusCallbackMethod'),
    'ConferenceRecordingStatusCallback': _.get(opts, 'conferenceRecordingStatusCallback'),
    'ConferenceRecordingStatusCallbackMethod': _.get(opts, 'conferenceRecordingStatusCallbackMethod'),
    'Region': _.get(opts, 'region'),
    'SipAuthUsername': _.get(opts, 'sipAuthUsername'),
    'SipAuthPassword': _.get(opts, 'sipAuthPassword'),
    'DequeueStatusCallbackEvent': serialize.map(_.get(opts, 'dequeueStatusCallbackEvent'), function(e) { return e; }),
    'PostWorkActivitySid': _.get(opts, 'postWorkActivitySid'),
    'SupervisorMode': _.get(opts, 'supervisorMode'),
    'Supervisor': _.get(opts, 'supervisor')
  });

  var promise = this._version.update({uri: this._uri, method: 'POST', data: data});

  promise = promise.then(function(payload) {
    deferred.resolve(new ReservationInstance(
      this._version,
      payload,
      this._solution.workspaceSid,
      this._solution.taskSid,
      this._solution.sid
    ));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

module.exports = {
  ReservationList: ReservationList,
  ReservationPage: ReservationPage,
  ReservationInstance: ReservationInstance,
  ReservationContext: ReservationContext
};
