// twilio/lib/rest/taskrouter/v1/workspace/worker/workersRealTimeStatistics.js
'use strict';

/* jshint ignore:start */
/**
 * This code was generated by
 * \ / _    _  _|   _  _
 *  | (_)\/(_)(_|\/| |(/_  v1.0.0
 *       /       /
 */
/* jshint ignore:end */

var Q = require('q');  /* jshint ignore:line */
var _ = require('lodash');  /* jshint ignore:line */
var Page = require('../../../../../base/Page');  /* jshint ignore:line */
var deserialize = require(
    '../../../../../base/deserialize');  /* jshint ignore:line */
var values = require('../../../../../base/values');  /* jshint ignore:line */

var WorkersRealTimeStatisticsList;
var WorkersRealTimeStatisticsPage;
var WorkersRealTimeStatisticsInstance;
var WorkersRealTimeStatisticsContext;

/* jshint ignore:start */
/**
 * @description Initialize the WorkersRealTimeStatisticsList
 *
 * @param {Twilio.Taskrouter.V1} version - Version of the resource
 * @param {string} workspaceSid - The workspace_sid
 */
/* jshint ignore:end */
WorkersRealTimeStatisticsList = function WorkersRealTimeStatisticsList(version,
    workspaceSid) {
  /* jshint ignore:start */
  /**
   * @param {string} sid - sid of instance
   *
   * @returns {Twilio.Taskrouter.V1.WorkspaceContext.WorkerContext.WorkersRealTimeStatisticsContext}
   */
  /* jshint ignore:end */
  function WorkersRealTimeStatisticsListInstance(sid) {
    return WorkersRealTimeStatisticsListInstance.get(sid);
  }

  WorkersRealTimeStatisticsListInstance._version = version;
  // Path Solution
  WorkersRealTimeStatisticsListInstance._solution = {workspaceSid: workspaceSid};
  /* jshint ignore:start */
  /**
   * Constructs a workers_real_time_statistics
   *
   * @returns {Twilio.Taskrouter.V1.WorkspaceContext.WorkerContext.WorkersRealTimeStatisticsContext}
   */
  /* jshint ignore:end */
  WorkersRealTimeStatisticsListInstance.get = function get() {
    return new WorkersRealTimeStatisticsContext(this._version, this._solution.workspaceSid);
  };

  return WorkersRealTimeStatisticsListInstance;
};


/* jshint ignore:start */
/**
 * Initialize the WorkersRealTimeStatisticsPage
 *
 * @param {V1} version - Version of the resource
 * @param {Response<string>} response - Response from the API
 * @param {WorkersRealTimeStatisticsSolution} solution - Path solution
 *
 * @returns WorkersRealTimeStatisticsPage
 */
/* jshint ignore:end */
WorkersRealTimeStatisticsPage = function WorkersRealTimeStatisticsPage(version,
    response, solution) {
  // Path Solution
  this._solution = solution;

  Page.prototype.constructor.call(this, version, response, this._solution);
};

_.extend(WorkersRealTimeStatisticsPage.prototype, Page.prototype);
WorkersRealTimeStatisticsPage.prototype.constructor = WorkersRealTimeStatisticsPage;

/* jshint ignore:start */
/**
 * Build an instance of WorkersRealTimeStatisticsInstance
 *
 * @param {WorkersRealTimeStatisticsPayload} payload -
 *          Payload response from the API
 *
 * @returns WorkersRealTimeStatisticsInstance
 */
/* jshint ignore:end */
WorkersRealTimeStatisticsPage.prototype.getInstance = function
    getInstance(payload) {
  return new WorkersRealTimeStatisticsInstance(this._version, payload, this._solution.workspaceSid);
};


/* jshint ignore:start */
/**
 * Initialize the WorkersRealTimeStatisticsContext
 *
 * @property {string} accountSid - The account_sid
 * @property {string} activityStatistics -
 *          The current Worker status count breakdown by Activity
 * @property {number} totalWorkers - The total number of Workers
 * @property {string} workspaceSid - The workspace_sid
 * @property {string} url - The url
 *
 * @param {V1} version - Version of the resource
 * @param {WorkersRealTimeStatisticsPayload} payload - The instance payload
 * @param {sid} workspaceSid - The workspace_sid
 */
/* jshint ignore:end */
WorkersRealTimeStatisticsInstance = function
    WorkersRealTimeStatisticsInstance(version, payload, workspaceSid) {
  this._version = version;

  // Marshaled Properties
  this.accountSid = payload.account_sid; // jshint ignore:line
  this.activityStatistics = payload.activity_statistics; // jshint ignore:line
  this.totalWorkers = deserialize.integer(payload.total_workers); // jshint ignore:line
  this.workspaceSid = payload.workspace_sid; // jshint ignore:line
  this.url = payload.url; // jshint ignore:line

  // Context
  this._context = undefined;
  this._solution = {workspaceSid: workspaceSid, };
};

Object.defineProperty(WorkersRealTimeStatisticsInstance.prototype,
  '_proxy', {
  get: function() {
    if (!this._context) {
      this._context = new WorkersRealTimeStatisticsContext(this._version, this._solution.workspaceSid);
    }

    return this._context;
  }
});

/* jshint ignore:start */
/**
 * fetch a WorkersRealTimeStatisticsInstance
 *
 * @param {object} [opts] - Options for request
 * @param {string} [opts.taskChannel] -
 *          Filter cumulative statistics by TaskChannel.
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed WorkersRealTimeStatisticsInstance
 */
/* jshint ignore:end */
WorkersRealTimeStatisticsInstance.prototype.fetch = function fetch(opts,
    callback) {
  return this._proxy.fetch(opts, callback);
};

/* jshint ignore:start */
/**
 * Produce a plain JSON object version of the WorkersRealTimeStatisticsInstance for serialization.
 * Removes any circular references in the object.
 *
 * @returns Object
 */
/* jshint ignore:end */
WorkersRealTimeStatisticsInstance.prototype.toJSON = function toJSON() {
  let clone = {};
  _.forOwn(this, function(value, key) {
    if (!_.startsWith(key, '_') && ! _.isFunction(value)) {
      clone[key] = value;
    }
  });
  return clone;
};


/* jshint ignore:start */
/**
 * Initialize the WorkersRealTimeStatisticsContext
 *
 * @param {V1} version - Version of the resource
 * @param {sid} workspaceSid - The workspace_sid
 */
/* jshint ignore:end */
WorkersRealTimeStatisticsContext = function
    WorkersRealTimeStatisticsContext(version, workspaceSid) {
  this._version = version;

  // Path Solution
  this._solution = {workspaceSid: workspaceSid, };
  this._uri = _.template(
    '/Workspaces/<%= workspaceSid %>/Workers/RealTimeStatistics' // jshint ignore:line
  )(this._solution);
};

/* jshint ignore:start */
/**
 * fetch a WorkersRealTimeStatisticsInstance
 *
 * @param {object} [opts] - Options for request
 * @param {string} [opts.taskChannel] -
 *          Filter cumulative statistics by TaskChannel.
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed WorkersRealTimeStatisticsInstance
 */
/* jshint ignore:end */
WorkersRealTimeStatisticsContext.prototype.fetch = function fetch(opts,
    callback) {
  if (_.isFunction(opts)) {
    callback = opts;
    opts = {};
  }
  opts = opts || {};

  var deferred = Q.defer();
  var data = values.of({'TaskChannel': _.get(opts, 'taskChannel')});

  var promise = this._version.fetch({uri: this._uri, method: 'GET', params: data});

  promise = promise.then(function(payload) {
    deferred.resolve(new WorkersRealTimeStatisticsInstance(
      this._version,
      payload,
      this._solution.workspaceSid
    ));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

module.exports = {
  WorkersRealTimeStatisticsList: WorkersRealTimeStatisticsList,
  WorkersRealTimeStatisticsPage: WorkersRealTimeStatisticsPage,
  WorkersRealTimeStatisticsInstance: WorkersRealTimeStatisticsInstance,
  WorkersRealTimeStatisticsContext: WorkersRealTimeStatisticsContext
};
