// twilio/lib/rest/video/v1/recordingSettings.js
'use strict';

/* jshint ignore:start */
/**
 * This code was generated by
 * \ / _    _  _|   _  _
 *  | (_)\/(_)(_|\/| |(/_  v1.0.0
 *       /       /
 */
/* jshint ignore:end */

var Q = require('q');  /* jshint ignore:line */
var _ = require('lodash');  /* jshint ignore:line */
var Page = require('../../../base/Page');  /* jshint ignore:line */
var serialize = require('../../../base/serialize');  /* jshint ignore:line */
var values = require('../../../base/values');  /* jshint ignore:line */

var RecordingSettingsList;
var RecordingSettingsPage;
var RecordingSettingsInstance;
var RecordingSettingsContext;

/* jshint ignore:start */
/**
 * @description Initialize the RecordingSettingsList
 * PLEASE NOTE that this class contains preview products that are subject to change. Use them with caution. If you currently do not have developer preview access, please contact help@twilio.com.
 *
 * @param {Twilio.Video.V1} version - Version of the resource
 */
/* jshint ignore:end */
RecordingSettingsList = function RecordingSettingsList(version) {
  /* jshint ignore:start */
  /**
   * @param {string} sid - sid of instance
   *
   * @returns {Twilio.Video.V1.RecordingSettingsContext}
   */
  /* jshint ignore:end */
  function RecordingSettingsListInstance(sid) {
    return RecordingSettingsListInstance.get(sid);
  }

  RecordingSettingsListInstance._version = version;
  // Path Solution
  RecordingSettingsListInstance._solution = {};
  /* jshint ignore:start */
  /**
   * Constructs a recording_settings
   *
   * @returns {Twilio.Video.V1.RecordingSettingsContext}
   */
  /* jshint ignore:end */
  RecordingSettingsListInstance.get = function get() {
    return new RecordingSettingsContext(this._version);
  };

  return RecordingSettingsListInstance;
};


/* jshint ignore:start */
/**
 * Initialize the RecordingSettingsPagePLEASE NOTE that this class contains preview products that are subject to change. Use them with caution. If you currently do not have developer preview access, please contact help@twilio.com.
 *
 * @param {V1} version - Version of the resource
 * @param {Response<string>} response - Response from the API
 * @param {RecordingSettingsSolution} solution - Path solution
 *
 * @returns RecordingSettingsPage
 */
/* jshint ignore:end */
RecordingSettingsPage = function RecordingSettingsPage(version, response,
                                                        solution) {
  // Path Solution
  this._solution = solution;

  Page.prototype.constructor.call(this, version, response, this._solution);
};

_.extend(RecordingSettingsPage.prototype, Page.prototype);
RecordingSettingsPage.prototype.constructor = RecordingSettingsPage;

/* jshint ignore:start */
/**
 * Build an instance of RecordingSettingsInstance
 *
 * @param {RecordingSettingsPayload} payload - Payload response from the API
 *
 * @returns RecordingSettingsInstance
 */
/* jshint ignore:end */
RecordingSettingsPage.prototype.getInstance = function getInstance(payload) {
  return new RecordingSettingsInstance(this._version, payload);
};


/* jshint ignore:start */
/**
 * Initialize the RecordingSettingsContextPLEASE NOTE that this class contains preview products that are subject to change. Use them with caution. If you currently do not have developer preview access, please contact help@twilio.com.
 *
 * @property {string} accountSid - The Twilio Account SID associated with this item
 * @property {string} friendlyName -
 *          Friendly name of the configuration to be shown in the console
 * @property {string} awsCredentialsSid -
 *          SID of the Stored Credential resource CRxx
 * @property {string} awsS3Url -
 *          URL of the S3 bucket where the recordings should be stored. We only support DNS-compliant URLs like http://<my-bucket>.s3-<aws-region>.amazonaws.com/recordings, where recordings is the path where you want recordings to be stored.
 * @property {boolean} awsStorageEnabled -
 *          true|false When set to true, all Recordings will be written to the AwsS3Url specified above. When set to false, all Recordings will be stored in Twilio's cloud.
 * @property {string} encryptionKeySid - SID of the Public Key resource CRxx
 * @property {boolean} encryptionEnabled -
 *          true|false When set to true, all Recordings will be stored encrypted.
 * @property {string} url - The url
 *
 * @param {V1} version - Version of the resource
 * @param {RecordingSettingsPayload} payload - The instance payload
 */
/* jshint ignore:end */
RecordingSettingsInstance = function RecordingSettingsInstance(version, payload)
    {
  this._version = version;

  // Marshaled Properties
  this.accountSid = payload.account_sid; // jshint ignore:line
  this.friendlyName = payload.friendly_name; // jshint ignore:line
  this.awsCredentialsSid = payload.aws_credentials_sid; // jshint ignore:line
  this.awsS3Url = payload.aws_s3_url; // jshint ignore:line
  this.awsStorageEnabled = payload.aws_storage_enabled; // jshint ignore:line
  this.encryptionKeySid = payload.encryption_key_sid; // jshint ignore:line
  this.encryptionEnabled = payload.encryption_enabled; // jshint ignore:line
  this.url = payload.url; // jshint ignore:line

  // Context
  this._context = undefined;
  this._solution = {};
};

Object.defineProperty(RecordingSettingsInstance.prototype,
  '_proxy', {
  get: function() {
    if (!this._context) {
      this._context = new RecordingSettingsContext(this._version);
    }

    return this._context;
  }
});

/* jshint ignore:start */
/**
 * fetch a RecordingSettingsInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed RecordingSettingsInstance
 */
/* jshint ignore:end */
RecordingSettingsInstance.prototype.fetch = function fetch(callback) {
  return this._proxy.fetch(callback);
};

/* jshint ignore:start */
/**
 * create a RecordingSettingsInstance
 *
 * @param {object} opts - Options for request
 * @param {string} opts.friendlyName -
 *          Friendly name of the configuration to be shown in the console
 * @param {string} [opts.awsCredentialsSid] -
 *          SID of the Stored Credential resource CRxx
 * @param {string} [opts.encryptionKeySid] - SID of the Public Key resource CRxx
 * @param {string} [opts.awsS3Url] -
 *          Identity of the external location where the recordings should be stored. We only support DNS-compliant URLs like http://<my-bucket>.s3-<aws-region>.amazonaws.com/recordings, where recordings is the path where you want recordings to be stored.
 * @param {boolean} [opts.awsStorageEnabled] -
 *          true|false When set to true, all Recordings will be written to the AwsS3Url specified above. When set to false, all Recordings will be stored in Twilio's cloud.
 * @param {boolean} [opts.encryptionEnabled] -
 *          true|false When set to true, all Recordings will be stored encrypted.
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed RecordingSettingsInstance
 */
/* jshint ignore:end */
RecordingSettingsInstance.prototype.create = function create(opts, callback) {
  return this._proxy.create(opts, callback);
};

/* jshint ignore:start */
/**
 * Produce a plain JSON object version of the RecordingSettingsInstance for serialization.
 * Removes any circular references in the object.
 *
 * @returns Object
 */
/* jshint ignore:end */
RecordingSettingsInstance.prototype.toJSON = function toJSON() {
  let clone = {};
  _.forOwn(this, function(value, key) {
    if (!_.startsWith(key, '_') && ! _.isFunction(value)) {
      clone[key] = value;
    }
  });
  return clone;
};


/* jshint ignore:start */
/**
 * Initialize the RecordingSettingsContextPLEASE NOTE that this class contains preview products that are subject to change. Use them with caution. If you currently do not have developer preview access, please contact help@twilio.com.
 *
 * @param {V1} version - Version of the resource
 */
/* jshint ignore:end */
RecordingSettingsContext = function RecordingSettingsContext(version) {
  this._version = version;

  // Path Solution
  this._solution = {};
  this._uri = _.template(
    '/RecordingSettings/Default' // jshint ignore:line
  )(this._solution);
};

/* jshint ignore:start */
/**
 * fetch a RecordingSettingsInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed RecordingSettingsInstance
 */
/* jshint ignore:end */
RecordingSettingsContext.prototype.fetch = function fetch(callback) {
  var deferred = Q.defer();
  var promise = this._version.fetch({uri: this._uri, method: 'GET'});

  promise = promise.then(function(payload) {
    deferred.resolve(new RecordingSettingsInstance(this._version, payload));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

/* jshint ignore:start */
/**
 * create a RecordingSettingsInstance
 *
 * @param {object} opts - Options for request
 * @param {string} opts.friendlyName -
 *          Friendly name of the configuration to be shown in the console
 * @param {string} [opts.awsCredentialsSid] -
 *          SID of the Stored Credential resource CRxx
 * @param {string} [opts.encryptionKeySid] - SID of the Public Key resource CRxx
 * @param {string} [opts.awsS3Url] -
 *          Identity of the external location where the recordings should be stored. We only support DNS-compliant URLs like http://<my-bucket>.s3-<aws-region>.amazonaws.com/recordings, where recordings is the path where you want recordings to be stored.
 * @param {boolean} [opts.awsStorageEnabled] -
 *          true|false When set to true, all Recordings will be written to the AwsS3Url specified above. When set to false, all Recordings will be stored in Twilio's cloud.
 * @param {boolean} [opts.encryptionEnabled] -
 *          true|false When set to true, all Recordings will be stored encrypted.
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed RecordingSettingsInstance
 */
/* jshint ignore:end */
RecordingSettingsContext.prototype.create = function create(opts, callback) {
  if (_.isUndefined(opts)) {
    throw new Error('Required parameter "opts" missing.');
  }
  if (_.isUndefined(opts.friendlyName)) {
    throw new Error('Required parameter "opts.friendlyName" missing.');
  }

  var deferred = Q.defer();
  var data = values.of({
    'FriendlyName': _.get(opts, 'friendlyName'),
    'AwsCredentialsSid': _.get(opts, 'awsCredentialsSid'),
    'EncryptionKeySid': _.get(opts, 'encryptionKeySid'),
    'AwsS3Url': _.get(opts, 'awsS3Url'),
    'AwsStorageEnabled': serialize.bool(_.get(opts, 'awsStorageEnabled')),
    'EncryptionEnabled': serialize.bool(_.get(opts, 'encryptionEnabled'))
  });

  var promise = this._version.create({uri: this._uri, method: 'POST', data: data});

  promise = promise.then(function(payload) {
    deferred.resolve(new RecordingSettingsInstance(this._version, payload));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

module.exports = {
  RecordingSettingsList: RecordingSettingsList,
  RecordingSettingsPage: RecordingSettingsPage,
  RecordingSettingsInstance: RecordingSettingsInstance,
  RecordingSettingsContext: RecordingSettingsContext
};
