// twilio/lib/rest/notify/v1/service/notification.js
'use strict';

/* jshint ignore:start */
/**
 * This code was generated by
 * \ / _    _  _|   _  _
 *  | (_)\/(_)(_|\/| |(/_  v1.0.0
 *       /       /
 */
/* jshint ignore:end */

var Q = require('q');  /* jshint ignore:line */
var _ = require('lodash');  /* jshint ignore:line */
var Page = require('../../../../base/Page');  /* jshint ignore:line */
var deserialize = require(
    '../../../../base/deserialize');  /* jshint ignore:line */
var serialize = require('../../../../base/serialize');  /* jshint ignore:line */
var values = require('../../../../base/values');  /* jshint ignore:line */

var NotificationList;
var NotificationPage;
var NotificationInstance;

/* jshint ignore:start */
/**
 * @description Initialize the NotificationList
 * PLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @param {Twilio.Notify.V1} version - Version of the resource
 * @param {string} serviceSid - The service_sid
 */
/* jshint ignore:end */
NotificationList = function NotificationList(version, serviceSid) {
  /* jshint ignore:start */
  /**
   * @param {string} sid - sid of instance
   *
   * @returns {Twilio.Notify.V1.ServiceContext.NotificationContext}
   */
  /* jshint ignore:end */
  function NotificationListInstance(sid) {
    return NotificationListInstance.get(sid);
  }

  NotificationListInstance._version = version;
  // Path Solution
  NotificationListInstance._solution = {serviceSid: serviceSid};
  NotificationListInstance._uri = _.template(
    '/Services/<%= serviceSid %>/Notifications' // jshint ignore:line
  )(NotificationListInstance._solution);
  /* jshint ignore:start */
  /**
   * create a NotificationInstance
   *
   * @param {object} [opts] - Options for request
   * @param {string} [opts.body] - Indicates the notification body text.
   * @param {notification.priority} [opts.priority] -
   *          Two priorities defined: low and high.
   * @param {number} [opts.ttl] -
   *          This parameter specifies how long the notification is valid.
   * @param {string} [opts.title] - Indicates the notification title.
   * @param {string} [opts.sound] - Indicates a sound to be played.
   * @param {string} [opts.action] -
   *          Specifies the actions to be displayed for the notification.
   * @param {string} [opts.data] -
   *          This parameter specifies the custom key-value pairs of the notification's payload.
   * @param {string} [opts.apn] -
   *          APNS specific payload that overrides corresponding attributes in a generic payload for Bindings with the apn BindingType.
   * @param {string} [opts.gcm] -
   *          GCM specific payload that overrides corresponding attributes in generic payload for Bindings with gcm BindingType.
   * @param {string} [opts.sms] -
   *          SMS specific payload that overrides corresponding attributes in generic payload for Bindings with sms BindingType.
   * @param {string} [opts.facebookMessenger] -
   *          Messenger specific payload that overrides corresponding attributes in generic payload for Bindings with facebook-messenger BindingType.
   * @param {string} [opts.fcm] -
   *          FCM specific payload that overrides corresponding attributes in generic payload for Bindings with fcm BindingType.
   * @param {string|list} [opts.segment] - The segment
   * @param {string} [opts.alexa] - The alexa
   * @param {string|list} [opts.toBinding] -
   *          The destination address in a JSON object.
   * @param {string|list} [opts.identity] -
   *          Delivery will be attempted only to Bindings with an Identity in this list.
   * @param {string|list} [opts.tag] -
   *          Delivery will be attempted only to Bindings that have all of the Tags in this list.
   * @param {function} [callback] - Callback to handle processed record
   *
   * @returns {Promise} Resolves to processed NotificationInstance
   */
  /* jshint ignore:end */
  NotificationListInstance.create = function create(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};

    var deferred = Q.defer();
    var data = values.of({
      'Identity': serialize.map(_.get(opts, 'identity'), function(e) { return e; }),
      'Tag': serialize.map(_.get(opts, 'tag'), function(e) { return e; }),
      'Body': _.get(opts, 'body'),
      'Priority': _.get(opts, 'priority'),
      'Ttl': _.get(opts, 'ttl'),
      'Title': _.get(opts, 'title'),
      'Sound': _.get(opts, 'sound'),
      'Action': _.get(opts, 'action'),
      'Data': serialize.object(_.get(opts, 'data')),
      'Apn': serialize.object(_.get(opts, 'apn')),
      'Gcm': serialize.object(_.get(opts, 'gcm')),
      'Sms': serialize.object(_.get(opts, 'sms')),
      'FacebookMessenger': serialize.object(_.get(opts, 'facebookMessenger')),
      'Fcm': serialize.object(_.get(opts, 'fcm')),
      'Segment': serialize.map(_.get(opts, 'segment'), function(e) { return e; }),
      'Alexa': serialize.object(_.get(opts, 'alexa')),
      'ToBinding': serialize.map(_.get(opts, 'toBinding'), function(e) { return e; })
    });

    var promise = this._version.create({uri: this._uri, method: 'POST', data: data});

    promise = promise.then(function(payload) {
      deferred.resolve(new NotificationInstance(this._version, payload));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  return NotificationListInstance;
};


/* jshint ignore:start */
/**
 * Initialize the NotificationPagePLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @param {V1} version - Version of the resource
 * @param {Response<string>} response - Response from the API
 * @param {NotificationSolution} solution - Path solution
 *
 * @returns NotificationPage
 */
/* jshint ignore:end */
NotificationPage = function NotificationPage(version, response, solution) {
  // Path Solution
  this._solution = solution;

  Page.prototype.constructor.call(this, version, response, this._solution);
};

_.extend(NotificationPage.prototype, Page.prototype);
NotificationPage.prototype.constructor = NotificationPage;

/* jshint ignore:start */
/**
 * Build an instance of NotificationInstance
 *
 * @param {NotificationPayload} payload - Payload response from the API
 *
 * @returns NotificationInstance
 */
/* jshint ignore:end */
NotificationPage.prototype.getInstance = function getInstance(payload) {
  return new NotificationInstance(this._version, payload, this._solution.serviceSid);
};


/* jshint ignore:start */
/**
 * Initialize the NotificationContextPLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @property {string} sid - The sid
 * @property {string} accountSid - The account_sid
 * @property {string} serviceSid - The service_sid
 * @property {Date} dateCreated - The date_created
 * @property {string} identities - List of Identities.
 * @property {string} tags - List of Tags
 * @property {string} segments - The segments
 * @property {notification.priority} priority -
 *          Two priorities defined: low and high.
 * @property {number} ttl -
 *          This parameter specifies how long the notification is valid.
 * @property {string} title - Indicates the notification title.
 * @property {string} body - Indicates the notification body text.
 * @property {string} sound - Indicates a sound to be played.
 * @property {string} action -
 *          Specifies the actions to be displayed for the notification.
 * @property {string} data -
 *          This parameter specifies the custom key-value pairs of the notification's payload.
 * @property {string} apn -
 *          APNS specific payload that overrides corresponding attributes in a generic payload for Bindings with the apn BindingType.
 * @property {string} gcm -
 *          GCM specific payload that overrides corresponding attributes in generic payload for Bindings with gcm BindingType.
 * @property {string} fcm -
 *          FCM specific payload that overrides corresponding attributes in generic payload for Bindings with fcm BindingType.
 * @property {string} sms -
 *          SMS specific payload that overrides corresponding attributes in generic payload for Bindings with sms BindingType.
 * @property {string} facebookMessenger -
 *          Messenger specific payload that overrides corresponding attributes in generic payload for Bindings with facebook-messenger BindingType.
 * @property {string} alexa - The alexa
 *
 * @param {V1} version - Version of the resource
 * @param {NotificationPayload} payload - The instance payload
 * @param {sid} serviceSid - The service_sid
 */
/* jshint ignore:end */
NotificationInstance = function NotificationInstance(version, payload,
                                                      serviceSid) {
  this._version = version;

  // Marshaled Properties
  this.sid = payload.sid; // jshint ignore:line
  this.accountSid = payload.account_sid; // jshint ignore:line
  this.serviceSid = payload.service_sid; // jshint ignore:line
  this.dateCreated = deserialize.iso8601DateTime(payload.date_created); // jshint ignore:line
  this.identities = payload.identities; // jshint ignore:line
  this.tags = payload.tags; // jshint ignore:line
  this.segments = payload.segments; // jshint ignore:line
  this.priority = payload.priority; // jshint ignore:line
  this.ttl = deserialize.integer(payload.ttl); // jshint ignore:line
  this.title = payload.title; // jshint ignore:line
  this.body = payload.body; // jshint ignore:line
  this.sound = payload.sound; // jshint ignore:line
  this.action = payload.action; // jshint ignore:line
  this.data = payload.data; // jshint ignore:line
  this.apn = payload.apn; // jshint ignore:line
  this.gcm = payload.gcm; // jshint ignore:line
  this.fcm = payload.fcm; // jshint ignore:line
  this.sms = payload.sms; // jshint ignore:line
  this.facebookMessenger = payload.facebook_messenger; // jshint ignore:line
  this.alexa = payload.alexa; // jshint ignore:line

  // Context
  this._context = undefined;
  this._solution = {serviceSid: serviceSid, };
};

/* jshint ignore:start */
/**
 * Produce a plain JSON object version of the NotificationInstance for serialization.
 * Removes any circular references in the object.
 *
 * @returns Object
 */
/* jshint ignore:end */
NotificationInstance.prototype.toJSON = function toJSON() {
  let clone = {};
  _.forOwn(this, function(value, key) {
    if (!_.startsWith(key, '_') && ! _.isFunction(value)) {
      clone[key] = value;
    }
  });
  return clone;
};

module.exports = {
  NotificationList: NotificationList,
  NotificationPage: NotificationPage,
  NotificationInstance: NotificationInstance
};
