// twilio/lib/rest/messaging/v1/service.js
'use strict';

/* jshint ignore:start */
/**
 * This code was generated by
 * \ / _    _  _|   _  _
 *  | (_)\/(_)(_|\/| |(/_  v1.0.0
 *       /       /
 */
/* jshint ignore:end */

var Q = require('q');  /* jshint ignore:line */
var _ = require('lodash');  /* jshint ignore:line */
var AlphaSenderList = require('./service/alphaSender').AlphaSenderList;
var Page = require('../../../base/Page');  /* jshint ignore:line */
var PhoneNumberList = require('./service/phoneNumber').PhoneNumberList;
var ShortCodeList = require('./service/shortCode').ShortCodeList;
var deserialize = require(
    '../../../base/deserialize');  /* jshint ignore:line */
var serialize = require('../../../base/serialize');  /* jshint ignore:line */
var values = require('../../../base/values');  /* jshint ignore:line */

var ServiceList;
var ServicePage;
var ServiceInstance;
var ServiceContext;

/* jshint ignore:start */
/**
 * @description Initialize the ServiceList
 * PLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @param {Twilio.Messaging.V1} version - Version of the resource
 */
/* jshint ignore:end */
ServiceList = function ServiceList(version) {
  /* jshint ignore:start */
  /**
   * @param {string} sid - sid of instance
   *
   * @returns {Twilio.Messaging.V1.ServiceContext}
   */
  /* jshint ignore:end */
  function ServiceListInstance(sid) {
    return ServiceListInstance.get(sid);
  }

  ServiceListInstance._version = version;
  // Path Solution
  ServiceListInstance._solution = {};
  ServiceListInstance._uri = _.template(
    '/Services' // jshint ignore:line
  )(ServiceListInstance._solution);
  /* jshint ignore:start */
  /**
   * create a ServiceInstance
   *
   * @param {object} opts - Options for request
   * @param {string} opts.friendlyName -
   *          A human readable descriptive text for this resource, up to 64 characters.
   * @param {string} [opts.inboundRequestUrl] -
   *          A webhook request is made to the Inbound Request URL when a message is received by any phone number or shortcode associated to your Messaging Service.
   * @param {string} [opts.inboundMethod] -
   *          The HTTP method used when making requests to the Inbound Request URL.
   * @param {string} [opts.fallbackUrl] -
   *          A request is made to the Fallback URL if an error occurs with retrieving or executing the TwiML from you Inbound Request URL.
   * @param {string} [opts.fallbackMethod] -
   *          The HTTP method used when requesting the Fallback URL.
   * @param {string} [opts.statusCallback] -
   *          A webhook request is made to the Status Callback to pass status updates about your messages.
   * @param {boolean} [opts.stickySender] -
   *          Configuration to enable or disable Sticky Sender on your Service Instance.
   * @param {boolean} [opts.mmsConverter] -
   *          Configuration to enable or disable MMS Converter on your Service Instance.
   * @param {boolean} [opts.smartEncoding] -
   *          Configuration to enable or disable Smart Encoding.
   * @param {service.scan_message_content} [opts.scanMessageContent] -
   *          The scan_message_content
   * @param {boolean} [opts.fallbackToLongCode] -
   *          Configuration to enable or disable Fallback to Long Code.
   * @param {boolean} [opts.areaCodeGeomatch] -
   *          Configuration to enable or disable Area Code Geomatch.
   * @param {number} [opts.validityPeriod] -
   *          Configuration to set the validity period of all messages sent from your Service, in seconds.
   * @param {boolean} [opts.synchronousValidation] - The synchronous_validation
   * @param {function} [callback] - Callback to handle processed record
   *
   * @returns {Promise} Resolves to processed ServiceInstance
   */
  /* jshint ignore:end */
  ServiceListInstance.create = function create(opts, callback) {
    if (_.isUndefined(opts)) {
      throw new Error('Required parameter "opts" missing.');
    }
    if (_.isUndefined(opts.friendlyName)) {
      throw new Error('Required parameter "opts.friendlyName" missing.');
    }

    var deferred = Q.defer();
    var data = values.of({
      'FriendlyName': _.get(opts, 'friendlyName'),
      'InboundRequestUrl': _.get(opts, 'inboundRequestUrl'),
      'InboundMethod': _.get(opts, 'inboundMethod'),
      'FallbackUrl': _.get(opts, 'fallbackUrl'),
      'FallbackMethod': _.get(opts, 'fallbackMethod'),
      'StatusCallback': _.get(opts, 'statusCallback'),
      'StickySender': serialize.bool(_.get(opts, 'stickySender')),
      'MmsConverter': serialize.bool(_.get(opts, 'mmsConverter')),
      'SmartEncoding': serialize.bool(_.get(opts, 'smartEncoding')),
      'ScanMessageContent': _.get(opts, 'scanMessageContent'),
      'FallbackToLongCode': serialize.bool(_.get(opts, 'fallbackToLongCode')),
      'AreaCodeGeomatch': serialize.bool(_.get(opts, 'areaCodeGeomatch')),
      'ValidityPeriod': _.get(opts, 'validityPeriod'),
      'SynchronousValidation': serialize.bool(_.get(opts, 'synchronousValidation'))
    });

    var promise = this._version.create({uri: this._uri, method: 'POST', data: data});

    promise = promise.then(function(payload) {
      deferred.resolve(new ServiceInstance(this._version, payload, this._solution.sid));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Streams ServiceInstance records from the API.
   *
   * This operation lazily loads records as efficiently as possible until the limit
   * is reached.
   *
   * The results are passed into the callback function, so this operation is memory efficient.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         each() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no pageSize is defined but a limit is defined,
   *         each() will attempt to read the limit with the most efficient
   *         page size, i.e. min(limit, 1000)
   * @param {Function} [opts.callback] -
   *         Function to process each record. If this and a positional
   *         callback are passed, this one will be used
   * @param {Function} [opts.done] -
   *          Function to be called upon completion of streaming
   * @param {Function} [callback] - Function to process each record
   */
  /* jshint ignore:end */
  ServiceListInstance.each = function each(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    if (opts.callback) {
      callback = opts.callback;
    }
    if (_.isUndefined(callback)) {
      throw new Error('Callback function must be provided');
    }

    var done = false;
    var currentPage = 1;
    var currentResource = 0;
    var limits = this._version.readLimits({
      limit: opts.limit,
      pageSize: opts.pageSize
    });

    function onComplete(error) {
      done = true;
      if (_.isFunction(opts.done)) {
        opts.done(error);
      }
    }

    function fetchNextPage(fn) {
      var promise = fn();
      if (_.isUndefined(promise)) {
        onComplete();
        return;
      }

      promise.then(function(page) {
        _.each(page.instances, function(instance) {
          if (done || (!_.isUndefined(opts.limit) && currentResource >= opts.limit)) {
            done = true;
            return false;
          }

          currentResource++;
          callback(instance, onComplete);
        });

        if ((limits.pageLimit && limits.pageLimit <= currentPage)) {
          onComplete();
        } else if (!done) {
          currentPage++;
          fetchNextPage(_.bind(page.nextPage, page));
        }
      });

      promise.catch(onComplete);
    }

    fetchNextPage(_.bind(this.page, this, _.merge(opts, limits)));
  };

  /* jshint ignore:start */
  /**
   * Lists ServiceInstance records from the API as a list.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         list() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no page_size is defined but a limit is defined,
   *         list() will attempt to read the limit with the most
   *         efficient page size, i.e. min(limit, 1000)
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  ServiceListInstance.list = function list(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    var deferred = Q.defer();
    var allResources = [];
    opts.callback = function(resource, done) {
      allResources.push(resource);

      if (!_.isUndefined(opts.limit) && allResources.length === opts.limit) {
        done();
      }
    };

    opts.done = function(error) {
      if (_.isUndefined(error)) {
        deferred.resolve(allResources);
      } else {
        deferred.reject(error);
      }
    };

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    this.each(opts);
    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single page of ServiceInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {string} [opts.pageToken] - PageToken provided by the API
   * @param {number} [opts.pageNumber] -
   *          Page Number, this value is simply for client state
   * @param {number} [opts.pageSize] - Number of records to return, defaults to 50
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  ServiceListInstance.page = function page(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};

    var deferred = Q.defer();
    var data = values.of({
      'PageToken': opts.pageToken,
      'Page': opts.pageNumber,
      'PageSize': opts.pageSize
    });

    var promise = this._version.page({uri: this._uri, method: 'GET', params: data});

    promise = promise.then(function(payload) {
      deferred.resolve(new ServicePage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single target page of ServiceInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {string} [targetUrl] - API-generated URL for the requested results page
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  ServiceListInstance.getPage = function getPage(targetUrl, callback) {
    var deferred = Q.defer();

    var promise = this._version._domain.twilio.request({method: 'GET', uri: targetUrl});

    promise = promise.then(function(payload) {
      deferred.resolve(new ServicePage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Constructs a service
   *
   * @param {string} sid - The sid
   *
   * @returns {Twilio.Messaging.V1.ServiceContext}
   */
  /* jshint ignore:end */
  ServiceListInstance.get = function get(sid) {
    return new ServiceContext(this._version, sid);
  };

  return ServiceListInstance;
};


/* jshint ignore:start */
/**
 * Initialize the ServicePagePLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @param {V1} version - Version of the resource
 * @param {Response<string>} response - Response from the API
 * @param {ServiceSolution} solution - Path solution
 *
 * @returns ServicePage
 */
/* jshint ignore:end */
ServicePage = function ServicePage(version, response, solution) {
  // Path Solution
  this._solution = solution;

  Page.prototype.constructor.call(this, version, response, this._solution);
};

_.extend(ServicePage.prototype, Page.prototype);
ServicePage.prototype.constructor = ServicePage;

/* jshint ignore:start */
/**
 * Build an instance of ServiceInstance
 *
 * @param {ServicePayload} payload - Payload response from the API
 *
 * @returns ServiceInstance
 */
/* jshint ignore:end */
ServicePage.prototype.getInstance = function getInstance(payload) {
  return new ServiceInstance(this._version, payload);
};


/* jshint ignore:start */
/**
 * Initialize the ServiceContextPLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @property {string} sid - Unique 34 character ID of the Service.
 * @property {string} accountSid -
 *          Unique 34 character ID of the Account that created this Service.
 * @property {string} friendlyName -
 *          A human readable descriptive text for this resource, up to 64 characters.
 * @property {Date} dateCreated - The date that this resource was created.
 * @property {Date} dateUpdated - The date that this resource was last updated.
 * @property {string} inboundRequestUrl -
 *          The URL Twilio will make a webhook request to when a message is received by any phone number or short code in your Service.
 * @property {string} inboundMethod -
 *          The HTTP method Twilio will use when making requests to the Inbound Request URL.
 * @property {string} fallbackUrl -
 *          The URL that Twilio will request if an error occurs when retrieving or executing the TwiML from your Inbound Request URL.
 * @property {string} fallbackMethod -
 *          The HTTP method Twilio will use when making requests to the Fallback URL.
 * @property {string} statusCallback -
 *          The URL Twilio will make a webhook request to when passing you status updates about the delivery of your messages.
 * @property {boolean} stickySender -
 *          Configuration to enable or disable Sticky Sender on your Service instance.
 * @property {boolean} mmsConverter -
 *          Configuration to enable or disable MMS Converter for messages sent through your Service instance.
 * @property {boolean} smartEncoding -
 *          Configuration to enable or disable Smart Encoding for messages sent through your Service instance.
 * @property {service.scan_message_content} scanMessageContent -
 *          The scan_message_content
 * @property {boolean} fallbackToLongCode -
 *          Configuration to enable or disable Fallback to Long Code for messages sent through your Service instance.
 * @property {boolean} areaCodeGeomatch -
 *          Configuration to enable or disable Area Code Geomatch on your Service Instance.
 * @property {boolean} synchronousValidation - The synchronous_validation
 * @property {number} validityPeriod -
 *          The number of seconds all messages sent from your Service are valid for.
 * @property {string} url - The url
 * @property {string} links - The links
 *
 * @param {V1} version - Version of the resource
 * @param {ServicePayload} payload - The instance payload
 * @param {sid} sid - The sid
 */
/* jshint ignore:end */
ServiceInstance = function ServiceInstance(version, payload, sid) {
  this._version = version;

  // Marshaled Properties
  this.sid = payload.sid; // jshint ignore:line
  this.accountSid = payload.account_sid; // jshint ignore:line
  this.friendlyName = payload.friendly_name; // jshint ignore:line
  this.dateCreated = deserialize.iso8601DateTime(payload.date_created); // jshint ignore:line
  this.dateUpdated = deserialize.iso8601DateTime(payload.date_updated); // jshint ignore:line
  this.inboundRequestUrl = payload.inbound_request_url; // jshint ignore:line
  this.inboundMethod = payload.inbound_method; // jshint ignore:line
  this.fallbackUrl = payload.fallback_url; // jshint ignore:line
  this.fallbackMethod = payload.fallback_method; // jshint ignore:line
  this.statusCallback = payload.status_callback; // jshint ignore:line
  this.stickySender = payload.sticky_sender; // jshint ignore:line
  this.mmsConverter = payload.mms_converter; // jshint ignore:line
  this.smartEncoding = payload.smart_encoding; // jshint ignore:line
  this.scanMessageContent = payload.scan_message_content; // jshint ignore:line
  this.fallbackToLongCode = payload.fallback_to_long_code; // jshint ignore:line
  this.areaCodeGeomatch = payload.area_code_geomatch; // jshint ignore:line
  this.synchronousValidation = payload.synchronous_validation; // jshint ignore:line
  this.validityPeriod = deserialize.integer(payload.validity_period); // jshint ignore:line
  this.url = payload.url; // jshint ignore:line
  this.links = payload.links; // jshint ignore:line

  // Context
  this._context = undefined;
  this._solution = {sid: sid || this.sid, };
};

Object.defineProperty(ServiceInstance.prototype,
  '_proxy', {
  get: function() {
    if (!this._context) {
      this._context = new ServiceContext(this._version, this._solution.sid);
    }

    return this._context;
  }
});

/* jshint ignore:start */
/**
 * update a ServiceInstance
 *
 * @param {object} [opts] - Options for request
 * @param {string} [opts.friendlyName] - The friendly_name
 * @param {string} [opts.inboundRequestUrl] - The inbound_request_url
 * @param {string} [opts.inboundMethod] - The inbound_method
 * @param {string} [opts.fallbackUrl] - The fallback_url
 * @param {string} [opts.fallbackMethod] - The fallback_method
 * @param {string} [opts.statusCallback] - The status_callback
 * @param {boolean} [opts.stickySender] - The sticky_sender
 * @param {boolean} [opts.mmsConverter] - The mms_converter
 * @param {boolean} [opts.smartEncoding] - The smart_encoding
 * @param {service.scan_message_content} [opts.scanMessageContent] -
 *          The scan_message_content
 * @param {boolean} [opts.fallbackToLongCode] - The fallback_to_long_code
 * @param {boolean} [opts.areaCodeGeomatch] - The area_code_geomatch
 * @param {number} [opts.validityPeriod] - The validity_period
 * @param {boolean} [opts.synchronousValidation] - The synchronous_validation
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ServiceInstance
 */
/* jshint ignore:end */
ServiceInstance.prototype.update = function update(opts, callback) {
  return this._proxy.update(opts, callback);
};

/* jshint ignore:start */
/**
 * fetch a ServiceInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ServiceInstance
 */
/* jshint ignore:end */
ServiceInstance.prototype.fetch = function fetch(callback) {
  return this._proxy.fetch(callback);
};

/* jshint ignore:start */
/**
 * remove a ServiceInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ServiceInstance
 */
/* jshint ignore:end */
ServiceInstance.prototype.remove = function remove(callback) {
  return this._proxy.remove(callback);
};

/* jshint ignore:start */
/**
 * Access the phoneNumbers
 *
 * @returns {Twilio.Messaging.V1.ServiceContext.PhoneNumberList}
 */
/* jshint ignore:end */
ServiceInstance.prototype.phoneNumbers = function phoneNumbers() {
  return this._proxy.phoneNumbers;
};

/* jshint ignore:start */
/**
 * Access the shortCodes
 *
 * @returns {Twilio.Messaging.V1.ServiceContext.ShortCodeList}
 */
/* jshint ignore:end */
ServiceInstance.prototype.shortCodes = function shortCodes() {
  return this._proxy.shortCodes;
};

/* jshint ignore:start */
/**
 * Access the alphaSenders
 *
 * @returns {Twilio.Messaging.V1.ServiceContext.AlphaSenderList}
 */
/* jshint ignore:end */
ServiceInstance.prototype.alphaSenders = function alphaSenders() {
  return this._proxy.alphaSenders;
};

/* jshint ignore:start */
/**
 * Produce a plain JSON object version of the ServiceInstance for serialization.
 * Removes any circular references in the object.
 *
 * @returns Object
 */
/* jshint ignore:end */
ServiceInstance.prototype.toJSON = function toJSON() {
  let clone = {};
  _.forOwn(this, function(value, key) {
    if (!_.startsWith(key, '_') && ! _.isFunction(value)) {
      clone[key] = value;
    }
  });
  return clone;
};


/* jshint ignore:start */
/**
 * Initialize the ServiceContextPLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @property {Twilio.Messaging.V1.ServiceContext.PhoneNumberList} phoneNumbers -
 *          phoneNumbers resource
 * @property {Twilio.Messaging.V1.ServiceContext.ShortCodeList} shortCodes -
 *          shortCodes resource
 * @property {Twilio.Messaging.V1.ServiceContext.AlphaSenderList} alphaSenders -
 *          alphaSenders resource
 *
 * @param {V1} version - Version of the resource
 * @param {sid} sid - The sid
 */
/* jshint ignore:end */
ServiceContext = function ServiceContext(version, sid) {
  this._version = version;

  // Path Solution
  this._solution = {sid: sid, };
  this._uri = _.template(
    '/Services/<%= sid %>' // jshint ignore:line
  )(this._solution);

  // Dependents
  this._phoneNumbers = undefined;
  this._shortCodes = undefined;
  this._alphaSenders = undefined;
};

/* jshint ignore:start */
/**
 * update a ServiceInstance
 *
 * @param {object} [opts] - Options for request
 * @param {string} [opts.friendlyName] - The friendly_name
 * @param {string} [opts.inboundRequestUrl] - The inbound_request_url
 * @param {string} [opts.inboundMethod] - The inbound_method
 * @param {string} [opts.fallbackUrl] - The fallback_url
 * @param {string} [opts.fallbackMethod] - The fallback_method
 * @param {string} [opts.statusCallback] - The status_callback
 * @param {boolean} [opts.stickySender] - The sticky_sender
 * @param {boolean} [opts.mmsConverter] - The mms_converter
 * @param {boolean} [opts.smartEncoding] - The smart_encoding
 * @param {service.scan_message_content} [opts.scanMessageContent] -
 *          The scan_message_content
 * @param {boolean} [opts.fallbackToLongCode] - The fallback_to_long_code
 * @param {boolean} [opts.areaCodeGeomatch] - The area_code_geomatch
 * @param {number} [opts.validityPeriod] - The validity_period
 * @param {boolean} [opts.synchronousValidation] - The synchronous_validation
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ServiceInstance
 */
/* jshint ignore:end */
ServiceContext.prototype.update = function update(opts, callback) {
  if (_.isFunction(opts)) {
    callback = opts;
    opts = {};
  }
  opts = opts || {};

  var deferred = Q.defer();
  var data = values.of({
    'FriendlyName': _.get(opts, 'friendlyName'),
    'InboundRequestUrl': _.get(opts, 'inboundRequestUrl'),
    'InboundMethod': _.get(opts, 'inboundMethod'),
    'FallbackUrl': _.get(opts, 'fallbackUrl'),
    'FallbackMethod': _.get(opts, 'fallbackMethod'),
    'StatusCallback': _.get(opts, 'statusCallback'),
    'StickySender': serialize.bool(_.get(opts, 'stickySender')),
    'MmsConverter': serialize.bool(_.get(opts, 'mmsConverter')),
    'SmartEncoding': serialize.bool(_.get(opts, 'smartEncoding')),
    'ScanMessageContent': _.get(opts, 'scanMessageContent'),
    'FallbackToLongCode': serialize.bool(_.get(opts, 'fallbackToLongCode')),
    'AreaCodeGeomatch': serialize.bool(_.get(opts, 'areaCodeGeomatch')),
    'ValidityPeriod': _.get(opts, 'validityPeriod'),
    'SynchronousValidation': serialize.bool(_.get(opts, 'synchronousValidation'))
  });

  var promise = this._version.update({uri: this._uri, method: 'POST', data: data});

  promise = promise.then(function(payload) {
    deferred.resolve(new ServiceInstance(this._version, payload, this._solution.sid));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

/* jshint ignore:start */
/**
 * fetch a ServiceInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ServiceInstance
 */
/* jshint ignore:end */
ServiceContext.prototype.fetch = function fetch(callback) {
  var deferred = Q.defer();
  var promise = this._version.fetch({uri: this._uri, method: 'GET'});

  promise = promise.then(function(payload) {
    deferred.resolve(new ServiceInstance(this._version, payload, this._solution.sid));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

/* jshint ignore:start */
/**
 * remove a ServiceInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ServiceInstance
 */
/* jshint ignore:end */
ServiceContext.prototype.remove = function remove(callback) {
  var deferred = Q.defer();
  var promise = this._version.remove({uri: this._uri, method: 'DELETE'});

  promise = promise.then(function(payload) {
    deferred.resolve(payload);
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

Object.defineProperty(ServiceContext.prototype,
  'phoneNumbers', {
  get: function() {
    if (!this._phoneNumbers) {
      this._phoneNumbers = new PhoneNumberList(this._version, this._solution.sid);
    }
    return this._phoneNumbers;
  }
});

Object.defineProperty(ServiceContext.prototype,
  'shortCodes', {
  get: function() {
    if (!this._shortCodes) {
      this._shortCodes = new ShortCodeList(this._version, this._solution.sid);
    }
    return this._shortCodes;
  }
});

Object.defineProperty(ServiceContext.prototype,
  'alphaSenders', {
  get: function() {
    if (!this._alphaSenders) {
      this._alphaSenders = new AlphaSenderList(this._version, this._solution.sid);
    }
    return this._alphaSenders;
  }
});

module.exports = {
  ServiceList: ServiceList,
  ServicePage: ServicePage,
  ServiceInstance: ServiceInstance,
  ServiceContext: ServiceContext
};
