// twilio/lib/rest/sync/v1/service/document.js
'use strict';

/* jshint ignore:start */
/**
 * This code was generated by
 * \ / _    _  _|   _  _
 *  | (_)\/(_)(_|\/| |(/_  v1.0.0
 *       /       /
 */
/* jshint ignore:end */

var Q = require('q');  /* jshint ignore:line */
var _ = require('lodash');  /* jshint ignore:line */
var DocumentPermissionList = require(
    './document/documentPermission').DocumentPermissionList;
var Page = require('../../../../base/Page');  /* jshint ignore:line */
var deserialize = require(
    '../../../../base/deserialize');  /* jshint ignore:line */
var serialize = require('../../../../base/serialize');  /* jshint ignore:line */
var values = require('../../../../base/values');  /* jshint ignore:line */

var DocumentList;
var DocumentPage;
var DocumentInstance;
var DocumentContext;

/* jshint ignore:start */
/**
 * @description Initialize the DocumentList
 * PLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @param {Twilio.Sync.V1} version - Version of the resource
 * @param {string} serviceSid -
 *          The unique SID identifier of the Service Instance that hosts this Document.
 */
/* jshint ignore:end */
DocumentList = function DocumentList(version, serviceSid) {
  /* jshint ignore:start */
  /**
   * @param {string} sid - sid of instance
   *
   * @returns {Twilio.Sync.V1.ServiceContext.DocumentContext}
   */
  /* jshint ignore:end */
  function DocumentListInstance(sid) {
    return DocumentListInstance.get(sid);
  }

  DocumentListInstance._version = version;
  // Path Solution
  DocumentListInstance._solution = {serviceSid: serviceSid};
  DocumentListInstance._uri = _.template(
    '/Services/<%= serviceSid %>/Documents' // jshint ignore:line
  )(DocumentListInstance._solution);
  /* jshint ignore:start */
  /**
   * create a DocumentInstance
   *
   * @param {object} [opts] - Options for request
   * @param {string} [opts.uniqueName] - Human-readable name for this document
   * @param {string} [opts.data] - JSON data to be stored in this document
   * @param {number} [opts.ttl] -
   *          Time-to-live of this Document in seconds, defaults to no expiration.
   * @param {function} [callback] - Callback to handle processed record
   *
   * @returns {Promise} Resolves to processed DocumentInstance
   */
  /* jshint ignore:end */
  DocumentListInstance.create = function create(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};

    var deferred = Q.defer();
    var data = values.of({
      'UniqueName': _.get(opts, 'uniqueName'),
      'Data': serialize.object(_.get(opts, 'data')),
      'Ttl': _.get(opts, 'ttl')
    });

    var promise = this._version.create({uri: this._uri, method: 'POST', data: data});

    promise = promise.then(function(payload) {
      deferred.resolve(new DocumentInstance(
        this._version,
        payload,
        this._solution.serviceSid,
        this._solution.sid
      ));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Streams DocumentInstance records from the API.
   *
   * This operation lazily loads records as efficiently as possible until the limit
   * is reached.
   *
   * The results are passed into the callback function, so this operation is memory efficient.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         each() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no pageSize is defined but a limit is defined,
   *         each() will attempt to read the limit with the most efficient
   *         page size, i.e. min(limit, 1000)
   * @param {Function} [opts.callback] -
   *         Function to process each record. If this and a positional
   *         callback are passed, this one will be used
   * @param {Function} [opts.done] -
   *          Function to be called upon completion of streaming
   * @param {Function} [callback] - Function to process each record
   */
  /* jshint ignore:end */
  DocumentListInstance.each = function each(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    if (opts.callback) {
      callback = opts.callback;
    }
    if (_.isUndefined(callback)) {
      throw new Error('Callback function must be provided');
    }

    var done = false;
    var currentPage = 1;
    var currentResource = 0;
    var limits = this._version.readLimits({
      limit: opts.limit,
      pageSize: opts.pageSize
    });

    function onComplete(error) {
      done = true;
      if (_.isFunction(opts.done)) {
        opts.done(error);
      }
    }

    function fetchNextPage(fn) {
      var promise = fn();
      if (_.isUndefined(promise)) {
        onComplete();
        return;
      }

      promise.then(function(page) {
        _.each(page.instances, function(instance) {
          if (done || (!_.isUndefined(opts.limit) && currentResource >= opts.limit)) {
            done = true;
            return false;
          }

          currentResource++;
          callback(instance, onComplete);
        });

        if ((limits.pageLimit && limits.pageLimit <= currentPage)) {
          onComplete();
        } else if (!done) {
          currentPage++;
          fetchNextPage(_.bind(page.nextPage, page));
        }
      });

      promise.catch(onComplete);
    }

    fetchNextPage(_.bind(this.page, this, _.merge(opts, limits)));
  };

  /* jshint ignore:start */
  /**
   * Lists DocumentInstance records from the API as a list.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         list() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no page_size is defined but a limit is defined,
   *         list() will attempt to read the limit with the most
   *         efficient page size, i.e. min(limit, 1000)
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  DocumentListInstance.list = function list(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    var deferred = Q.defer();
    var allResources = [];
    opts.callback = function(resource, done) {
      allResources.push(resource);

      if (!_.isUndefined(opts.limit) && allResources.length === opts.limit) {
        done();
      }
    };

    opts.done = function(error) {
      if (_.isUndefined(error)) {
        deferred.resolve(allResources);
      } else {
        deferred.reject(error);
      }
    };

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    this.each(opts);
    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single page of DocumentInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {string} [opts.pageToken] - PageToken provided by the API
   * @param {number} [opts.pageNumber] -
   *          Page Number, this value is simply for client state
   * @param {number} [opts.pageSize] - Number of records to return, defaults to 50
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  DocumentListInstance.page = function page(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};

    var deferred = Q.defer();
    var data = values.of({
      'PageToken': opts.pageToken,
      'Page': opts.pageNumber,
      'PageSize': opts.pageSize
    });

    var promise = this._version.page({uri: this._uri, method: 'GET', params: data});

    promise = promise.then(function(payload) {
      deferred.resolve(new DocumentPage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single target page of DocumentInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {string} [targetUrl] - API-generated URL for the requested results page
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  DocumentListInstance.getPage = function getPage(targetUrl, callback) {
    var deferred = Q.defer();

    var promise = this._version._domain.twilio.request({method: 'GET', uri: targetUrl});

    promise = promise.then(function(payload) {
      deferred.resolve(new DocumentPage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Constructs a document
   *
   * @param {string} sid - The sid
   *
   * @returns {Twilio.Sync.V1.ServiceContext.DocumentContext}
   */
  /* jshint ignore:end */
  DocumentListInstance.get = function get(sid) {
    return new DocumentContext(this._version, this._solution.serviceSid, sid);
  };

  return DocumentListInstance;
};


/* jshint ignore:start */
/**
 * Initialize the DocumentPagePLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @param {V1} version - Version of the resource
 * @param {Response<string>} response - Response from the API
 * @param {DocumentSolution} solution - Path solution
 *
 * @returns DocumentPage
 */
/* jshint ignore:end */
DocumentPage = function DocumentPage(version, response, solution) {
  // Path Solution
  this._solution = solution;

  Page.prototype.constructor.call(this, version, response, this._solution);
};

_.extend(DocumentPage.prototype, Page.prototype);
DocumentPage.prototype.constructor = DocumentPage;

/* jshint ignore:start */
/**
 * Build an instance of DocumentInstance
 *
 * @param {DocumentPayload} payload - Payload response from the API
 *
 * @returns DocumentInstance
 */
/* jshint ignore:end */
DocumentPage.prototype.getInstance = function getInstance(payload) {
  return new DocumentInstance(this._version, payload, this._solution.serviceSid);
};


/* jshint ignore:start */
/**
 * Initialize the DocumentContextPLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @property {string} sid - The unique 34-character SID identifier of the Document.
 * @property {string} uniqueName -
 *          The unique and addressable name of this Document.
 * @property {string} accountSid - The unique SID identifier of the Twilio Account.
 * @property {string} serviceSid -
 *          The unique SID identifier of the Service Instance that hosts this Document.
 * @property {string} url - The absolute URL for this Document.
 * @property {string} links -
 *          A dictionary of URL links to nested resources of this Document.
 * @property {string} revision -
 *          Contains the current revision of this Document, represented by a string identifier.
 * @property {string} data -
 *          Contains arbitrary user-defined, schema-less data that this Document stores, represented by a JSON object, up to 16KB.
 * @property {Date} dateExpires -
 *          Contains the date this Document expires and gets deleted automatically.
 * @property {Date} dateCreated -
 *          The date this Document was created, given in UTC ISO 8601 format.
 * @property {Date} dateUpdated -
 *          Specifies the date this Document was last updated, given in UTC ISO 8601 format.
 * @property {string} createdBy - The identity of the Document creator.
 *
 * @param {V1} version - Version of the resource
 * @param {DocumentPayload} payload - The instance payload
 * @param {sid} serviceSid -
 *          The unique SID identifier of the Service Instance that hosts this Document.
 * @param {sid_like} sid - The sid
 */
/* jshint ignore:end */
DocumentInstance = function DocumentInstance(version, payload, serviceSid, sid)
                                              {
  this._version = version;

  // Marshaled Properties
  this.sid = payload.sid; // jshint ignore:line
  this.uniqueName = payload.unique_name; // jshint ignore:line
  this.accountSid = payload.account_sid; // jshint ignore:line
  this.serviceSid = payload.service_sid; // jshint ignore:line
  this.url = payload.url; // jshint ignore:line
  this.links = payload.links; // jshint ignore:line
  this.revision = payload.revision; // jshint ignore:line
  this.data = payload.data; // jshint ignore:line
  this.dateExpires = deserialize.iso8601DateTime(payload.date_expires); // jshint ignore:line
  this.dateCreated = deserialize.iso8601DateTime(payload.date_created); // jshint ignore:line
  this.dateUpdated = deserialize.iso8601DateTime(payload.date_updated); // jshint ignore:line
  this.createdBy = payload.created_by; // jshint ignore:line

  // Context
  this._context = undefined;
  this._solution = {serviceSid: serviceSid, sid: sid || this.sid, };
};

Object.defineProperty(DocumentInstance.prototype,
  '_proxy', {
  get: function() {
    if (!this._context) {
      this._context = new DocumentContext(this._version, this._solution.serviceSid, this._solution.sid);
    }

    return this._context;
  }
});

/* jshint ignore:start */
/**
 * fetch a DocumentInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed DocumentInstance
 */
/* jshint ignore:end */
DocumentInstance.prototype.fetch = function fetch(callback) {
  return this._proxy.fetch(callback);
};

/* jshint ignore:start */
/**
 * remove a DocumentInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed DocumentInstance
 */
/* jshint ignore:end */
DocumentInstance.prototype.remove = function remove(callback) {
  return this._proxy.remove(callback);
};

/* jshint ignore:start */
/**
 * update a DocumentInstance
 *
 * @param {object} [opts] - Options for request
 * @param {string} [opts.data] -
 *          Contains an arbitrary JSON object to be stored in this Document.
 * @param {number} [opts.ttl] - New time-to-live of this Document in seconds.
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed DocumentInstance
 */
/* jshint ignore:end */
DocumentInstance.prototype.update = function update(opts, callback) {
  return this._proxy.update(opts, callback);
};

/* jshint ignore:start */
/**
 * Access the documentPermissions
 *
 * @returns {Twilio.Sync.V1.ServiceContext.DocumentContext.DocumentPermissionList}
 */
/* jshint ignore:end */
DocumentInstance.prototype.documentPermissions = function documentPermissions()
    {
  return this._proxy.documentPermissions;
};

/* jshint ignore:start */
/**
 * Produce a plain JSON object version of the DocumentInstance for serialization.
 * Removes any circular references in the object.
 *
 * @returns Object
 */
/* jshint ignore:end */
DocumentInstance.prototype.toJSON = function toJSON() {
  let clone = {};
  _.forOwn(this, function(value, key) {
    if (!_.startsWith(key, '_') && ! _.isFunction(value)) {
      clone[key] = value;
    }
  });
  return clone;
};


/* jshint ignore:start */
/**
 * Initialize the DocumentContextPLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @property {Twilio.Sync.V1.ServiceContext.DocumentContext.DocumentPermissionList} documentPermissions -
 *          documentPermissions resource
 *
 * @param {V1} version - Version of the resource
 * @param {sid_like} serviceSid - The service_sid
 * @param {sid_like} sid - The sid
 */
/* jshint ignore:end */
DocumentContext = function DocumentContext(version, serviceSid, sid) {
  this._version = version;

  // Path Solution
  this._solution = {serviceSid: serviceSid, sid: sid, };
  this._uri = _.template(
    '/Services/<%= serviceSid %>/Documents/<%= sid %>' // jshint ignore:line
  )(this._solution);

  // Dependents
  this._documentPermissions = undefined;
};

/* jshint ignore:start */
/**
 * fetch a DocumentInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed DocumentInstance
 */
/* jshint ignore:end */
DocumentContext.prototype.fetch = function fetch(callback) {
  var deferred = Q.defer();
  var promise = this._version.fetch({uri: this._uri, method: 'GET'});

  promise = promise.then(function(payload) {
    deferred.resolve(new DocumentInstance(
      this._version,
      payload,
      this._solution.serviceSid,
      this._solution.sid
    ));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

/* jshint ignore:start */
/**
 * remove a DocumentInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed DocumentInstance
 */
/* jshint ignore:end */
DocumentContext.prototype.remove = function remove(callback) {
  var deferred = Q.defer();
  var promise = this._version.remove({uri: this._uri, method: 'DELETE'});

  promise = promise.then(function(payload) {
    deferred.resolve(payload);
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

/* jshint ignore:start */
/**
 * update a DocumentInstance
 *
 * @param {object} [opts] - Options for request
 * @param {string} [opts.data] -
 *          Contains an arbitrary JSON object to be stored in this Document.
 * @param {number} [opts.ttl] - New time-to-live of this Document in seconds.
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed DocumentInstance
 */
/* jshint ignore:end */
DocumentContext.prototype.update = function update(opts, callback) {
  if (_.isFunction(opts)) {
    callback = opts;
    opts = {};
  }
  opts = opts || {};

  var deferred = Q.defer();
  var data = values.of({'Data': serialize.object(_.get(opts, 'data')), 'Ttl': _.get(opts, 'ttl')});

  var promise = this._version.update({uri: this._uri, method: 'POST', data: data});

  promise = promise.then(function(payload) {
    deferred.resolve(new DocumentInstance(
      this._version,
      payload,
      this._solution.serviceSid,
      this._solution.sid
    ));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

Object.defineProperty(DocumentContext.prototype,
  'documentPermissions', {
  get: function() {
    if (!this._documentPermissions) {
      this._documentPermissions = new DocumentPermissionList(
        this._version,
        this._solution.serviceSid,
        this._solution.sid
      );
    }
    return this._documentPermissions;
  }
});

module.exports = {
  DocumentList: DocumentList,
  DocumentPage: DocumentPage,
  DocumentInstance: DocumentInstance,
  DocumentContext: DocumentContext
};
