// twilio/lib/rest/studio/v1/flow/execution/executionStep/executionStepContext.js
'use strict';

/* jshint ignore:start */
/**
 * This code was generated by
 * \ / _    _  _|   _  _
 *  | (_)\/(_)(_|\/| |(/_  v1.0.0
 *       /       /
 */
/* jshint ignore:end */

var Q = require('q');  /* jshint ignore:line */
var _ = require('lodash');  /* jshint ignore:line */
var Page = require('../../../../../../base/Page');  /* jshint ignore:line */
var values = require('../../../../../../base/values');  /* jshint ignore:line */

var ExecutionStepContextList;
var ExecutionStepContextPage;
var ExecutionStepContextInstance;
var ExecutionStepContextContext;

/* jshint ignore:start */
/**
 * @description Initialize the ExecutionStepContextList
 *
 * @param {Twilio.Studio.V1} version - Version of the resource
 * @param {string} flowSid - Flow Sid.
 * @param {string} executionSid - Execution Sid.
 * @param {string} stepSid - Step Sid.
 */
/* jshint ignore:end */
ExecutionStepContextList = function ExecutionStepContextList(version, flowSid,
    executionSid, stepSid) {
  /* jshint ignore:start */
  /**
   * @param {string} sid - sid of instance
   *
   * @returns {Twilio.Studio.V1.FlowContext.ExecutionContext.ExecutionStepContext.ExecutionStepContextContext}
   */
  /* jshint ignore:end */
  function ExecutionStepContextListInstance(sid) {
    return ExecutionStepContextListInstance.get(sid);
  }

  ExecutionStepContextListInstance._version = version;
  // Path Solution
  ExecutionStepContextListInstance._solution = {
    flowSid: flowSid,
    executionSid: executionSid,
    stepSid: stepSid
  };
  /* jshint ignore:start */
  /**
   * Constructs a execution_step_context
   *
   * @returns {Twilio.Studio.V1.FlowContext.ExecutionContext.ExecutionStepContext.ExecutionStepContextContext}
   */
  /* jshint ignore:end */
  ExecutionStepContextListInstance.get = function get() {
    return new ExecutionStepContextContext(
      this._version,
      this._solution.flowSid,
      this._solution.executionSid,
      this._solution.stepSid
    );
  };

  return ExecutionStepContextListInstance;
};


/* jshint ignore:start */
/**
 * Initialize the ExecutionStepContextPage
 *
 * @param {V1} version - Version of the resource
 * @param {Response<string>} response - Response from the API
 * @param {ExecutionStepContextSolution} solution - Path solution
 *
 * @returns ExecutionStepContextPage
 */
/* jshint ignore:end */
ExecutionStepContextPage = function ExecutionStepContextPage(version, response,
    solution) {
  // Path Solution
  this._solution = solution;

  Page.prototype.constructor.call(this, version, response, this._solution);
};

_.extend(ExecutionStepContextPage.prototype, Page.prototype);
ExecutionStepContextPage.prototype.constructor = ExecutionStepContextPage;

/* jshint ignore:start */
/**
 * Build an instance of ExecutionStepContextInstance
 *
 * @param {ExecutionStepContextPayload} payload - Payload response from the API
 *
 * @returns ExecutionStepContextInstance
 */
/* jshint ignore:end */
ExecutionStepContextPage.prototype.getInstance = function getInstance(payload) {
  return new ExecutionStepContextInstance(
    this._version,
    payload,
    this._solution.flowSid,
    this._solution.executionSid,
    this._solution.stepSid
  );
};


/* jshint ignore:start */
/**
 * Initialize the ExecutionStepContextContext
 *
 * @property {string} accountSid - Account Sid.
 * @property {string} context - Flow state.
 * @property {string} executionSid - Execution Sid.
 * @property {string} flowSid - Flow Sid.
 * @property {string} stepSid - Step Sid.
 * @property {string} url - The URL of this resource.
 *
 * @param {V1} version - Version of the resource
 * @param {ExecutionStepContextPayload} payload - The instance payload
 * @param {sid} flowSid - Flow Sid.
 * @param {sid} executionSid - Execution Sid.
 * @param {sid} stepSid - Step Sid.
 */
/* jshint ignore:end */
ExecutionStepContextInstance = function ExecutionStepContextInstance(version,
    payload, flowSid, executionSid, stepSid) {
  this._version = version;

  // Marshaled Properties
  this.accountSid = payload.account_sid; // jshint ignore:line
  this.context = payload.context; // jshint ignore:line
  this.executionSid = payload.execution_sid; // jshint ignore:line
  this.flowSid = payload.flow_sid; // jshint ignore:line
  this.stepSid = payload.step_sid; // jshint ignore:line
  this.url = payload.url; // jshint ignore:line

  // Context
  this._context = undefined;
  this._solution = {flowSid: flowSid, executionSid: executionSid, stepSid: stepSid, };
};

Object.defineProperty(ExecutionStepContextInstance.prototype,
  '_proxy', {
  get: function() {
    if (!this._context) {
      this._context = new ExecutionStepContextContext(
        this._version,
        this._solution.flowSid,
        this._solution.executionSid,
        this._solution.stepSid
      );
    }

    return this._context;
  }
});

/* jshint ignore:start */
/**
 * fetch a ExecutionStepContextInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ExecutionStepContextInstance
 */
/* jshint ignore:end */
ExecutionStepContextInstance.prototype.fetch = function fetch(callback) {
  return this._proxy.fetch(callback);
};

/* jshint ignore:start */
/**
 * Produce a plain JSON object version of the ExecutionStepContextInstance for serialization.
 * Removes any circular references in the object.
 *
 * @returns Object
 */
/* jshint ignore:end */
ExecutionStepContextInstance.prototype.toJSON = function toJSON() {
  let clone = {};
  _.forOwn(this, function(value, key) {
    if (!_.startsWith(key, '_') && ! _.isFunction(value)) {
      clone[key] = value;
    }
  });
  return clone;
};


/* jshint ignore:start */
/**
 * Initialize the ExecutionStepContextContext
 *
 * @param {V1} version - Version of the resource
 * @param {sid} flowSid - Flow Sid.
 * @param {sid} executionSid - Execution Sid.
 * @param {sid} stepSid - Step Sid.
 */
/* jshint ignore:end */
ExecutionStepContextContext = function ExecutionStepContextContext(version,
    flowSid, executionSid, stepSid) {
  this._version = version;

  // Path Solution
  this._solution = {flowSid: flowSid, executionSid: executionSid, stepSid: stepSid, };
  this._uri = _.template(
    '/Flows/<%= flowSid %>/Executions/<%= executionSid %>/Steps/<%= stepSid %>/Context' // jshint ignore:line
  )(this._solution);
};

/* jshint ignore:start */
/**
 * fetch a ExecutionStepContextInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed ExecutionStepContextInstance
 */
/* jshint ignore:end */
ExecutionStepContextContext.prototype.fetch = function fetch(callback) {
  var deferred = Q.defer();
  var promise = this._version.fetch({uri: this._uri, method: 'GET'});

  promise = promise.then(function(payload) {
    deferred.resolve(new ExecutionStepContextInstance(
      this._version,
      payload,
      this._solution.flowSid,
      this._solution.executionSid,
      this._solution.stepSid
    ));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

module.exports = {
  ExecutionStepContextList: ExecutionStepContextList,
  ExecutionStepContextPage: ExecutionStepContextPage,
  ExecutionStepContextInstance: ExecutionStepContextInstance,
  ExecutionStepContextContext: ExecutionStepContextContext
};
