// twilio/lib/rest/preview/marketplace/availableAddOn.js
'use strict';

/* jshint ignore:start */
/**
 * This code was generated by
 * \ / _    _  _|   _  _
 *  | (_)\/(_)(_|\/| |(/_  v1.0.0
 *       /       /
 */
/* jshint ignore:end */

var Q = require('q');  /* jshint ignore:line */
var _ = require('lodash');  /* jshint ignore:line */
var AvailableAddOnExtensionList = require(
    './availableAddOn/availableAddOnExtension').AvailableAddOnExtensionList;
var Page = require('../../../base/Page');  /* jshint ignore:line */
var values = require('../../../base/values');  /* jshint ignore:line */

var AvailableAddOnList;
var AvailableAddOnPage;
var AvailableAddOnInstance;
var AvailableAddOnContext;

/* jshint ignore:start */
/**
 * @description Initialize the AvailableAddOnList
 * PLEASE NOTE that this class contains preview products that are subject to change. Use them with caution. If you currently do not have developer preview access, please contact help@twilio.com.
 *
 * @param {Twilio.Preview.Marketplace} version - Version of the resource
 */
/* jshint ignore:end */
AvailableAddOnList = function AvailableAddOnList(version) {
  /* jshint ignore:start */
  /**
   * @param {string} sid - sid of instance
   *
   * @returns {Twilio.Preview.Marketplace.AvailableAddOnContext}
   */
  /* jshint ignore:end */
  function AvailableAddOnListInstance(sid) {
    return AvailableAddOnListInstance.get(sid);
  }

  AvailableAddOnListInstance._version = version;
  // Path Solution
  AvailableAddOnListInstance._solution = {};
  AvailableAddOnListInstance._uri = _.template(
    '/AvailableAddOns' // jshint ignore:line
  )(AvailableAddOnListInstance._solution);
  /* jshint ignore:start */
  /**
   * Streams AvailableAddOnInstance records from the API.
   *
   * This operation lazily loads records as efficiently as possible until the limit
   * is reached.
   *
   * The results are passed into the callback function, so this operation is memory efficient.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         each() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no pageSize is defined but a limit is defined,
   *         each() will attempt to read the limit with the most efficient
   *         page size, i.e. min(limit, 1000)
   * @param {Function} [opts.callback] -
   *         Function to process each record. If this and a positional
   *         callback are passed, this one will be used
   * @param {Function} [opts.done] -
   *          Function to be called upon completion of streaming
   * @param {Function} [callback] - Function to process each record
   */
  /* jshint ignore:end */
  AvailableAddOnListInstance.each = function each(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    if (opts.callback) {
      callback = opts.callback;
    }
    if (_.isUndefined(callback)) {
      throw new Error('Callback function must be provided');
    }

    var done = false;
    var currentPage = 1;
    var currentResource = 0;
    var limits = this._version.readLimits({
      limit: opts.limit,
      pageSize: opts.pageSize
    });

    function onComplete(error) {
      done = true;
      if (_.isFunction(opts.done)) {
        opts.done(error);
      }
    }

    function fetchNextPage(fn) {
      var promise = fn();
      if (_.isUndefined(promise)) {
        onComplete();
        return;
      }

      promise.then(function(page) {
        _.each(page.instances, function(instance) {
          if (done || (!_.isUndefined(opts.limit) && currentResource >= opts.limit)) {
            done = true;
            return false;
          }

          currentResource++;
          callback(instance, onComplete);
        });

        if ((limits.pageLimit && limits.pageLimit <= currentPage)) {
          onComplete();
        } else if (!done) {
          currentPage++;
          fetchNextPage(_.bind(page.nextPage, page));
        }
      });

      promise.catch(onComplete);
    }

    fetchNextPage(_.bind(this.page, this, _.merge(opts, limits)));
  };

  /* jshint ignore:start */
  /**
   * Lists AvailableAddOnInstance records from the API as a list.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         list() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no page_size is defined but a limit is defined,
   *         list() will attempt to read the limit with the most
   *         efficient page size, i.e. min(limit, 1000)
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  AvailableAddOnListInstance.list = function list(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    var deferred = Q.defer();
    var allResources = [];
    opts.callback = function(resource, done) {
      allResources.push(resource);

      if (!_.isUndefined(opts.limit) && allResources.length === opts.limit) {
        done();
      }
    };

    opts.done = function(error) {
      if (_.isUndefined(error)) {
        deferred.resolve(allResources);
      } else {
        deferred.reject(error);
      }
    };

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    this.each(opts);
    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single page of AvailableAddOnInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {string} [opts.pageToken] - PageToken provided by the API
   * @param {number} [opts.pageNumber] -
   *          Page Number, this value is simply for client state
   * @param {number} [opts.pageSize] - Number of records to return, defaults to 50
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  AvailableAddOnListInstance.page = function page(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};

    var deferred = Q.defer();
    var data = values.of({
      'PageToken': opts.pageToken,
      'Page': opts.pageNumber,
      'PageSize': opts.pageSize
    });

    var promise = this._version.page({uri: this._uri, method: 'GET', params: data});

    promise = promise.then(function(payload) {
      deferred.resolve(new AvailableAddOnPage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single target page of AvailableAddOnInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {string} [targetUrl] - API-generated URL for the requested results page
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  AvailableAddOnListInstance.getPage = function getPage(targetUrl, callback) {
    var deferred = Q.defer();

    var promise = this._version._domain.twilio.request({method: 'GET', uri: targetUrl});

    promise = promise.then(function(payload) {
      deferred.resolve(new AvailableAddOnPage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Constructs a available_add_on
   *
   * @param {string} sid - The unique Available Add-on Sid
   *
   * @returns {Twilio.Preview.Marketplace.AvailableAddOnContext}
   */
  /* jshint ignore:end */
  AvailableAddOnListInstance.get = function get(sid) {
    return new AvailableAddOnContext(this._version, sid);
  };

  return AvailableAddOnListInstance;
};


/* jshint ignore:start */
/**
 * Initialize the AvailableAddOnPagePLEASE NOTE that this class contains preview products that are subject to change. Use them with caution. If you currently do not have developer preview access, please contact help@twilio.com.
 *
 * @param {Marketplace} version - Version of the resource
 * @param {Response<string>} response - Response from the API
 * @param {AvailableAddOnSolution} solution - Path solution
 *
 * @returns AvailableAddOnPage
 */
/* jshint ignore:end */
AvailableAddOnPage = function AvailableAddOnPage(version, response, solution) {
  // Path Solution
  this._solution = solution;

  Page.prototype.constructor.call(this, version, response, this._solution);
};

_.extend(AvailableAddOnPage.prototype, Page.prototype);
AvailableAddOnPage.prototype.constructor = AvailableAddOnPage;

/* jshint ignore:start */
/**
 * Build an instance of AvailableAddOnInstance
 *
 * @param {AvailableAddOnPayload} payload - Payload response from the API
 *
 * @returns AvailableAddOnInstance
 */
/* jshint ignore:end */
AvailableAddOnPage.prototype.getInstance = function getInstance(payload) {
  return new AvailableAddOnInstance(this._version, payload);
};


/* jshint ignore:start */
/**
 * Initialize the AvailableAddOnContextPLEASE NOTE that this class contains preview products that are subject to change. Use them with caution. If you currently do not have developer preview access, please contact help@twilio.com.
 *
 * @property {string} sid - A string that uniquely identifies this Add-on
 * @property {string} friendlyName - A description of this Add-on
 * @property {string} description - A short description of the Add-on functionality
 * @property {string} pricingType -
 *          The way customers are charged for using this Add-on
 * @property {string} configurationSchema -
 *          The JSON Schema describing the Add-on's configuration
 * @property {string} url - The url
 * @property {string} links - The links
 *
 * @param {Marketplace} version - Version of the resource
 * @param {AvailableAddOnPayload} payload - The instance payload
 * @param {sid} sid - The unique Available Add-on Sid
 */
/* jshint ignore:end */
AvailableAddOnInstance = function AvailableAddOnInstance(version, payload, sid)
                                                          {
  this._version = version;

  // Marshaled Properties
  this.sid = payload.sid; // jshint ignore:line
  this.friendlyName = payload.friendly_name; // jshint ignore:line
  this.description = payload.description; // jshint ignore:line
  this.pricingType = payload.pricing_type; // jshint ignore:line
  this.configurationSchema = payload.configuration_schema; // jshint ignore:line
  this.url = payload.url; // jshint ignore:line
  this.links = payload.links; // jshint ignore:line

  // Context
  this._context = undefined;
  this._solution = {sid: sid || this.sid, };
};

Object.defineProperty(AvailableAddOnInstance.prototype,
  '_proxy', {
  get: function() {
    if (!this._context) {
      this._context = new AvailableAddOnContext(this._version, this._solution.sid);
    }

    return this._context;
  }
});

/* jshint ignore:start */
/**
 * fetch a AvailableAddOnInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed AvailableAddOnInstance
 */
/* jshint ignore:end */
AvailableAddOnInstance.prototype.fetch = function fetch(callback) {
  return this._proxy.fetch(callback);
};

/* jshint ignore:start */
/**
 * Access the extensions
 *
 * @returns {Twilio.Preview.Marketplace.AvailableAddOnContext.AvailableAddOnExtensionList}
 */
/* jshint ignore:end */
AvailableAddOnInstance.prototype.extensions = function extensions() {
  return this._proxy.extensions;
};

/* jshint ignore:start */
/**
 * Produce a plain JSON object version of the AvailableAddOnInstance for serialization.
 * Removes any circular references in the object.
 *
 * @returns Object
 */
/* jshint ignore:end */
AvailableAddOnInstance.prototype.toJSON = function toJSON() {
  let clone = {};
  _.forOwn(this, function(value, key) {
    if (!_.startsWith(key, '_') && ! _.isFunction(value)) {
      clone[key] = value;
    }
  });
  return clone;
};


/* jshint ignore:start */
/**
 * Initialize the AvailableAddOnContextPLEASE NOTE that this class contains preview products that are subject to change. Use them with caution. If you currently do not have developer preview access, please contact help@twilio.com.
 *
 * @property {Twilio.Preview.Marketplace.AvailableAddOnContext.AvailableAddOnExtensionList} extensions -
 *          extensions resource
 *
 * @param {Marketplace} version - Version of the resource
 * @param {sid} sid - The unique Available Add-on Sid
 */
/* jshint ignore:end */
AvailableAddOnContext = function AvailableAddOnContext(version, sid) {
  this._version = version;

  // Path Solution
  this._solution = {sid: sid, };
  this._uri = _.template(
    '/AvailableAddOns/<%= sid %>' // jshint ignore:line
  )(this._solution);

  // Dependents
  this._extensions = undefined;
};

/* jshint ignore:start */
/**
 * fetch a AvailableAddOnInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed AvailableAddOnInstance
 */
/* jshint ignore:end */
AvailableAddOnContext.prototype.fetch = function fetch(callback) {
  var deferred = Q.defer();
  var promise = this._version.fetch({uri: this._uri, method: 'GET'});

  promise = promise.then(function(payload) {
    deferred.resolve(new AvailableAddOnInstance(this._version, payload, this._solution.sid));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

Object.defineProperty(AvailableAddOnContext.prototype,
  'extensions', {
  get: function() {
    if (!this._extensions) {
      this._extensions = new AvailableAddOnExtensionList(this._version, this._solution.sid);
    }
    return this._extensions;
  }
});

module.exports = {
  AvailableAddOnList: AvailableAddOnList,
  AvailableAddOnPage: AvailableAddOnPage,
  AvailableAddOnInstance: AvailableAddOnInstance,
  AvailableAddOnContext: AvailableAddOnContext
};
