// twilio/lib/rest/sync/v1/service/syncMap/syncMapItem.js
'use strict';

/* jshint ignore:start */
/**
 * This code was generated by
 * \ / _    _  _|   _  _
 *  | (_)\/(_)(_|\/| |(/_  v1.0.0
 *       /       /
 */
/* jshint ignore:end */

var Q = require('q');  /* jshint ignore:line */
var _ = require('lodash');  /* jshint ignore:line */
var Page = require('../../../../../base/Page');  /* jshint ignore:line */
var deserialize = require(
    '../../../../../base/deserialize');  /* jshint ignore:line */
var serialize = require(
    '../../../../../base/serialize');  /* jshint ignore:line */
var values = require('../../../../../base/values');  /* jshint ignore:line */

var SyncMapItemList;
var SyncMapItemPage;
var SyncMapItemInstance;
var SyncMapItemContext;

/* jshint ignore:start */
/**
 * @description Initialize the SyncMapItemList
 * PLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @param {Twilio.Sync.V1} version - Version of the resource
 * @param {string} serviceSid -
 *          The unique SID identifier of the Service Instance that hosts this Map object.
 * @param {string} mapSid -
 *          The unique 34-character SID identifier of the Map containing this Item.
 */
/* jshint ignore:end */
SyncMapItemList = function SyncMapItemList(version, serviceSid, mapSid) {
  /* jshint ignore:start */
  /**
   * @param {string} sid - sid of instance
   *
   * @returns {Twilio.Sync.V1.ServiceContext.SyncMapContext.SyncMapItemContext}
   */
  /* jshint ignore:end */
  function SyncMapItemListInstance(sid) {
    return SyncMapItemListInstance.get(sid);
  }

  SyncMapItemListInstance._version = version;
  // Path Solution
  SyncMapItemListInstance._solution = {serviceSid: serviceSid, mapSid: mapSid};
  SyncMapItemListInstance._uri = _.template(
    '/Services/<%= serviceSid %>/Maps/<%= mapSid %>/Items' // jshint ignore:line
  )(SyncMapItemListInstance._solution);
  /* jshint ignore:start */
  /**
   * create a SyncMapItemInstance
   *
   * @param {object} opts - Options for request
   * @param {string} opts.key - The unique user-defined key of this Map Item.
   * @param {string} opts.data -
   *          Contains arbitrary user-defined, schema-less data that this Map Item stores, represented by a JSON object, up to 16KB.
   * @param {number} [opts.ttl] -
   *          Time-to-live of this Map in seconds, defaults to no expiration.
   * @param {function} [callback] - Callback to handle processed record
   *
   * @returns {Promise} Resolves to processed SyncMapItemInstance
   */
  /* jshint ignore:end */
  SyncMapItemListInstance.create = function create(opts, callback) {
    if (_.isUndefined(opts)) {
      throw new Error('Required parameter "opts" missing.');
    }
    if (_.isUndefined(opts.key)) {
      throw new Error('Required parameter "opts.key" missing.');
    }
    if (_.isUndefined(opts.data)) {
      throw new Error('Required parameter "opts.data" missing.');
    }

    var deferred = Q.defer();
    var data = values.of({
      'Key': _.get(opts, 'key'),
      'Data': serialize.object(_.get(opts, 'data')),
      'Ttl': _.get(opts, 'ttl')
    });

    var promise = this._version.create({uri: this._uri, method: 'POST', data: data});

    promise = promise.then(function(payload) {
      deferred.resolve(new SyncMapItemInstance(
        this._version,
        payload,
        this._solution.serviceSid,
        this._solution.mapSid,
        this._solution.key
      ));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Streams SyncMapItemInstance records from the API.
   *
   * This operation lazily loads records as efficiently as possible until the limit
   * is reached.
   *
   * The results are passed into the callback function, so this operation is memory efficient.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {sync_map_item.query_result_order} [opts.order] -
   *          A string; asc or desc. Map Items are ordered lexicographically by Item key.
   * @param {string} [opts.from] - The Item key offset (including the specified key).
   * @param {sync_map_item.query_from_bound_type} [opts.bounds] - The bounds
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         each() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no pageSize is defined but a limit is defined,
   *         each() will attempt to read the limit with the most efficient
   *         page size, i.e. min(limit, 1000)
   * @param {Function} [opts.callback] -
   *         Function to process each record. If this and a positional
   *         callback are passed, this one will be used
   * @param {Function} [opts.done] -
   *          Function to be called upon completion of streaming
   * @param {Function} [callback] - Function to process each record
   */
  /* jshint ignore:end */
  SyncMapItemListInstance.each = function each(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    if (opts.callback) {
      callback = opts.callback;
    }
    if (_.isUndefined(callback)) {
      throw new Error('Callback function must be provided');
    }

    var done = false;
    var currentPage = 1;
    var currentResource = 0;
    var limits = this._version.readLimits({
      limit: opts.limit,
      pageSize: opts.pageSize
    });

    function onComplete(error) {
      done = true;
      if (_.isFunction(opts.done)) {
        opts.done(error);
      }
    }

    function fetchNextPage(fn) {
      var promise = fn();
      if (_.isUndefined(promise)) {
        onComplete();
        return;
      }

      promise.then(function(page) {
        _.each(page.instances, function(instance) {
          if (done || (!_.isUndefined(opts.limit) && currentResource >= opts.limit)) {
            done = true;
            return false;
          }

          currentResource++;
          callback(instance, onComplete);
        });

        if ((limits.pageLimit && limits.pageLimit <= currentPage)) {
          onComplete();
        } else if (!done) {
          currentPage++;
          fetchNextPage(_.bind(page.nextPage, page));
        }
      });

      promise.catch(onComplete);
    }

    fetchNextPage(_.bind(this.page, this, _.merge(opts, limits)));
  };

  /* jshint ignore:start */
  /**
   * Lists SyncMapItemInstance records from the API as a list.
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {sync_map_item.query_result_order} [opts.order] -
   *          A string; asc or desc. Map Items are ordered lexicographically by Item key.
   * @param {string} [opts.from] - The Item key offset (including the specified key).
   * @param {sync_map_item.query_from_bound_type} [opts.bounds] - The bounds
   * @param {number} [opts.limit] -
   *         Upper limit for the number of records to return.
   *         list() guarantees never to return more than limit.
   *         Default is no limit
   * @param {number} [opts.pageSize] -
   *         Number of records to fetch per request,
   *         when not set will use the default value of 50 records.
   *         If no page_size is defined but a limit is defined,
   *         list() will attempt to read the limit with the most
   *         efficient page size, i.e. min(limit, 1000)
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  SyncMapItemListInstance.list = function list(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};
    var deferred = Q.defer();
    var allResources = [];
    opts.callback = function(resource, done) {
      allResources.push(resource);

      if (!_.isUndefined(opts.limit) && allResources.length === opts.limit) {
        done();
      }
    };

    opts.done = function(error) {
      if (_.isUndefined(error)) {
        deferred.resolve(allResources);
      } else {
        deferred.reject(error);
      }
    };

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    this.each(opts);
    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single page of SyncMapItemInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {object} [opts] - Options for request
   * @param {sync_map_item.query_result_order} [opts.order] -
   *          A string; asc or desc. Map Items are ordered lexicographically by Item key.
   * @param {string} [opts.from] - The Item key offset (including the specified key).
   * @param {sync_map_item.query_from_bound_type} [opts.bounds] - The bounds
   * @param {string} [opts.pageToken] - PageToken provided by the API
   * @param {number} [opts.pageNumber] -
   *          Page Number, this value is simply for client state
   * @param {number} [opts.pageSize] - Number of records to return, defaults to 50
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  SyncMapItemListInstance.page = function page(opts, callback) {
    if (_.isFunction(opts)) {
      callback = opts;
      opts = {};
    }
    opts = opts || {};

    var deferred = Q.defer();
    var data = values.of({
      'Order': _.get(opts, 'order'),
      'From': _.get(opts, 'from'),
      'Bounds': _.get(opts, 'bounds'),
      'PageToken': opts.pageToken,
      'Page': opts.pageNumber,
      'PageSize': opts.pageSize
    });

    var promise = this._version.page({uri: this._uri, method: 'GET', params: data});

    promise = promise.then(function(payload) {
      deferred.resolve(new SyncMapItemPage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Retrieve a single target page of SyncMapItemInstance records from the API.
   * Request is executed immediately
   *
   * If a function is passed as the first argument, it will be used as the callback function.
   *
   * @param {string} [targetUrl] - API-generated URL for the requested results page
   * @param {function} [callback] - Callback to handle list of records
   *
   * @returns {Promise} Resolves to a list of records
   */
  /* jshint ignore:end */
  SyncMapItemListInstance.getPage = function getPage(targetUrl, callback) {
    var deferred = Q.defer();

    var promise = this._version._domain.twilio.request({method: 'GET', uri: targetUrl});

    promise = promise.then(function(payload) {
      deferred.resolve(new SyncMapItemPage(this._version, payload, this._solution));
    }.bind(this));

    promise.catch(function(error) {
      deferred.reject(error);
    });

    if (_.isFunction(callback)) {
      deferred.promise.nodeify(callback);
    }

    return deferred.promise;
  };

  /* jshint ignore:start */
  /**
   * Constructs a sync_map_item
   *
   * @param {string} key - The key
   *
   * @returns {Twilio.Sync.V1.ServiceContext.SyncMapContext.SyncMapItemContext}
   */
  /* jshint ignore:end */
  SyncMapItemListInstance.get = function get(key) {
    return new SyncMapItemContext(this._version, this._solution.serviceSid, this._solution.mapSid, key);
  };

  return SyncMapItemListInstance;
};


/* jshint ignore:start */
/**
 * Initialize the SyncMapItemPagePLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @param {V1} version - Version of the resource
 * @param {Response<string>} response - Response from the API
 * @param {SyncMapItemSolution} solution - Path solution
 *
 * @returns SyncMapItemPage
 */
/* jshint ignore:end */
SyncMapItemPage = function SyncMapItemPage(version, response, solution) {
  // Path Solution
  this._solution = solution;

  Page.prototype.constructor.call(this, version, response, this._solution);
};

_.extend(SyncMapItemPage.prototype, Page.prototype);
SyncMapItemPage.prototype.constructor = SyncMapItemPage;

/* jshint ignore:start */
/**
 * Build an instance of SyncMapItemInstance
 *
 * @param {SyncMapItemPayload} payload - Payload response from the API
 *
 * @returns SyncMapItemInstance
 */
/* jshint ignore:end */
SyncMapItemPage.prototype.getInstance = function getInstance(payload) {
  return new SyncMapItemInstance(
    this._version,
    payload,
    this._solution.serviceSid,
    this._solution.mapSid
  );
};


/* jshint ignore:start */
/**
 * Initialize the SyncMapItemContextPLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @property {string} key - The unique user-defined key of this Map Item.
 * @property {string} accountSid - The unique SID identifier of the Twilio Account.
 * @property {string} serviceSid -
 *          The unique SID identifier of the Service Instance that hosts this Map object.
 * @property {string} mapSid -
 *          The unique 34-character SID identifier of the Map containing this Item.
 * @property {string} url - The absolute URL for this Map.
 * @property {string} revision -
 *          Contains the current revision of this Map, represented by a string identifier.
 * @property {string} data -
 *          Contains arbitrary user-defined, schema-less data that this Map Item stores, represented by a JSON object, up to 16KB.
 * @property {Date} dateExpires -
 *          Contains the date this Map expires and gets deleted automatically.
 * @property {Date} dateCreated -
 *          The date this Map was created, given in UTC ISO 8601 format.
 * @property {Date} dateUpdated -
 *          Specifies the date this Map was last updated, given in UTC ISO 8601 format.
 * @property {string} createdBy - The identity of the Map creator.
 *
 * @param {V1} version - Version of the resource
 * @param {SyncMapItemPayload} payload - The instance payload
 * @param {sid} serviceSid -
 *          The unique SID identifier of the Service Instance that hosts this Map object.
 * @param {sid} mapSid -
 *          The unique 34-character SID identifier of the Map containing this Item.
 * @param {string} key - The key
 */
/* jshint ignore:end */
SyncMapItemInstance = function SyncMapItemInstance(version, payload, serviceSid,
                                                    mapSid, key) {
  this._version = version;

  // Marshaled Properties
  this.key = payload.key; // jshint ignore:line
  this.accountSid = payload.account_sid; // jshint ignore:line
  this.serviceSid = payload.service_sid; // jshint ignore:line
  this.mapSid = payload.map_sid; // jshint ignore:line
  this.url = payload.url; // jshint ignore:line
  this.revision = payload.revision; // jshint ignore:line
  this.data = payload.data; // jshint ignore:line
  this.dateExpires = deserialize.iso8601DateTime(payload.date_expires); // jshint ignore:line
  this.dateCreated = deserialize.iso8601DateTime(payload.date_created); // jshint ignore:line
  this.dateUpdated = deserialize.iso8601DateTime(payload.date_updated); // jshint ignore:line
  this.createdBy = payload.created_by; // jshint ignore:line

  // Context
  this._context = undefined;
  this._solution = {serviceSid: serviceSid, mapSid: mapSid, key: key || this.key, };
};

Object.defineProperty(SyncMapItemInstance.prototype,
  '_proxy', {
  get: function() {
    if (!this._context) {
      this._context = new SyncMapItemContext(
        this._version,
        this._solution.serviceSid,
        this._solution.mapSid,
        this._solution.key
      );
    }

    return this._context;
  }
});

/* jshint ignore:start */
/**
 * fetch a SyncMapItemInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed SyncMapItemInstance
 */
/* jshint ignore:end */
SyncMapItemInstance.prototype.fetch = function fetch(callback) {
  return this._proxy.fetch(callback);
};

/* jshint ignore:start */
/**
 * remove a SyncMapItemInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed SyncMapItemInstance
 */
/* jshint ignore:end */
SyncMapItemInstance.prototype.remove = function remove(callback) {
  return this._proxy.remove(callback);
};

/* jshint ignore:start */
/**
 * update a SyncMapItemInstance
 *
 * @param {object} [opts] - Options for request
 * @param {string} [opts.data] -
 *          Contains an arbitrary JSON object to be stored in this Map Item.
 * @param {number} [opts.ttl] - New time-to-live of this Map in seconds.
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed SyncMapItemInstance
 */
/* jshint ignore:end */
SyncMapItemInstance.prototype.update = function update(opts, callback) {
  return this._proxy.update(opts, callback);
};

/* jshint ignore:start */
/**
 * Produce a plain JSON object version of the SyncMapItemInstance for serialization.
 * Removes any circular references in the object.
 *
 * @returns Object
 */
/* jshint ignore:end */
SyncMapItemInstance.prototype.toJSON = function toJSON() {
  let clone = {};
  _.forOwn(this, function(value, key) {
    if (!_.startsWith(key, '_') && ! _.isFunction(value)) {
      clone[key] = value;
    }
  });
  return clone;
};


/* jshint ignore:start */
/**
 * Initialize the SyncMapItemContextPLEASE NOTE that this class contains beta products that are subject to change. Use them with caution.
 *
 * @param {V1} version - Version of the resource
 * @param {sid_like} serviceSid - The service_sid
 * @param {sid_like} mapSid - The map_sid
 * @param {string} key - The key
 */
/* jshint ignore:end */
SyncMapItemContext = function SyncMapItemContext(version, serviceSid, mapSid,
                                                  key) {
  this._version = version;

  // Path Solution
  this._solution = {serviceSid: serviceSid, mapSid: mapSid, key: key, };
  this._uri = _.template(
    '/Services/<%= serviceSid %>/Maps/<%= mapSid %>/Items/<%= key %>' // jshint ignore:line
  )(this._solution);
};

/* jshint ignore:start */
/**
 * fetch a SyncMapItemInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed SyncMapItemInstance
 */
/* jshint ignore:end */
SyncMapItemContext.prototype.fetch = function fetch(callback) {
  var deferred = Q.defer();
  var promise = this._version.fetch({uri: this._uri, method: 'GET'});

  promise = promise.then(function(payload) {
    deferred.resolve(new SyncMapItemInstance(
      this._version,
      payload,
      this._solution.serviceSid,
      this._solution.mapSid,
      this._solution.key
    ));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

/* jshint ignore:start */
/**
 * remove a SyncMapItemInstance
 *
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed SyncMapItemInstance
 */
/* jshint ignore:end */
SyncMapItemContext.prototype.remove = function remove(callback) {
  var deferred = Q.defer();
  var promise = this._version.remove({uri: this._uri, method: 'DELETE'});

  promise = promise.then(function(payload) {
    deferred.resolve(payload);
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

/* jshint ignore:start */
/**
 * update a SyncMapItemInstance
 *
 * @param {object} [opts] - Options for request
 * @param {string} [opts.data] -
 *          Contains an arbitrary JSON object to be stored in this Map Item.
 * @param {number} [opts.ttl] - New time-to-live of this Map in seconds.
 * @param {function} [callback] - Callback to handle processed record
 *
 * @returns {Promise} Resolves to processed SyncMapItemInstance
 */
/* jshint ignore:end */
SyncMapItemContext.prototype.update = function update(opts, callback) {
  if (_.isFunction(opts)) {
    callback = opts;
    opts = {};
  }
  opts = opts || {};

  var deferred = Q.defer();
  var data = values.of({'Data': serialize.object(_.get(opts, 'data')), 'Ttl': _.get(opts, 'ttl')});

  var promise = this._version.update({uri: this._uri, method: 'POST', data: data});

  promise = promise.then(function(payload) {
    deferred.resolve(new SyncMapItemInstance(
      this._version,
      payload,
      this._solution.serviceSid,
      this._solution.mapSid,
      this._solution.key
    ));
  }.bind(this));

  promise.catch(function(error) {
    deferred.reject(error);
  });

  if (_.isFunction(callback)) {
    deferred.promise.nodeify(callback);
  }

  return deferred.promise;
};

module.exports = {
  SyncMapItemList: SyncMapItemList,
  SyncMapItemPage: SyncMapItemPage,
  SyncMapItemInstance: SyncMapItemInstance,
  SyncMapItemContext: SyncMapItemContext
};
